{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "dN4hIpS2yb39"
   },
   "source": [
    "Project 2. SARS-CoV\n",
    "\n",
    "You are in charge of a public health office in a large metropolitan area. There’s currently a COVID-19 outbreak, and due to recent budget cuts, you’ve also been appointed as the new bioinformatics expert (congratulations on the promotion!). Your responsibilities now include developing a pipeline to analyze NGS sequencing data for genome assembly, mutation detection, and lineage identification of sequenced samples.\n",
    "\n",
    "You have two genomes to work with, and your pipeline should be capable of identifying mutations in both.\n",
    "\n",
    "To achieve full credit (100%), submit the following deliverables:\n",
    "\n",
    "1) Two tables (DataFrames) listing mutations for each genome.\n",
    "\n",
    "2) A third table showing mutations shared between both genomes.\n",
    "\n",
    "3) A simple chart (using any of the chart types introduced in the previous module) that visualizes an aspect of the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "bXEZOk7nydjB",
    "outputId": "c30ab01d-79c6-422b-e596-6325533e5485"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading...\n",
      "From (original): https://drive.google.com/uc?id=1-2h_NXGEvUIcSgtQnpeRy3BN8QKmbARf\n",
      "From (redirected): https://drive.google.com/uc?id=1-2h_NXGEvUIcSgtQnpeRy3BN8QKmbARf&confirm=t&uuid=d9edfd66-1ccf-4c39-b6de-bebf7bdc3d52\n",
      "To: /content/SARS.R2.fastq\n",
      "100% 154M/154M [00:02<00:00, 71.1MB/s]\n",
      "Downloading...\n",
      "From (original): https://drive.google.com/uc?id=1-1zFQh9lo7O1Fe6dR6abgwqBZlu4TBB-\n",
      "From (redirected): https://drive.google.com/uc?id=1-1zFQh9lo7O1Fe6dR6abgwqBZlu4TBB-&confirm=t&uuid=8e3aa078-4a50-4222-a750-cf6685faa7a2\n",
      "To: /content/SARS.R1.fastq\n",
      "100% 154M/154M [00:01<00:00, 135MB/s]\n",
      "Downloading...\n",
      "From (original): https://drive.google.com/uc?id=1-3y7YBrnJmpwdxTqimDGewWpTIkzlWei\n",
      "From (redirected): https://drive.google.com/uc?id=1-3y7YBrnJmpwdxTqimDGewWpTIkzlWei&confirm=t&uuid=6e28d4ba-1248-4fc8-9592-5566d86a977c\n",
      "To: /content/SARS-LONG-EDT.R2.fastq\n",
      "100% 189M/189M [00:01<00:00, 116MB/s]\n",
      "Downloading...\n",
      "From (original): https://drive.google.com/uc?id=1-2jBX4sV28m7FoauSMH2Tik65h4m72ig\n",
      "From (redirected): https://drive.google.com/uc?id=1-2jBX4sV28m7FoauSMH2Tik65h4m72ig&confirm=t&uuid=372f5f1e-0d85-4ec1-b1b4-13dba56ef5f8\n",
      "To: /content/SARS-LONG-EDT.R1.fastq\n",
      "100% 189M/189M [00:01<00:00, 188MB/s]\n"
     ]
    }
   ],
   "source": [
    "#download genomes\n",
    "!gdown 1-2h_NXGEvUIcSgtQnpeRy3BN8QKmbARf\n",
    "!gdown 1-1zFQh9lo7O1Fe6dR6abgwqBZlu4TBB-\n",
    "\n",
    "!gdown 1-3y7YBrnJmpwdxTqimDGewWpTIkzlWei\n",
    "!gdown 1-2jBX4sV28m7FoauSMH2Tik65h4m72ig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "czK_cHok4v60",
    "outputId": "8a8a4fa2-0817-4ab1-bf7b-e51c0d187129"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading...\n",
      "From: https://drive.google.com/uc?id=1yTOH-9w0RbcBhwetNWzYxFXENtiyPlzW\n",
      "To: /content/WuhanRef.fasta\n",
      "\r  0% 0.00/30.4k [00:00<?, ?B/s]\r100% 30.4k/30.4k [00:00<00:00, 48.7MB/s]\n"
     ]
    }
   ],
   "source": [
    "#wuhan reference genome\n",
    "!gdown 1yTOH-9w0RbcBhwetNWzYxFXENtiyPlzW"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "collapsed": true,
    "id": "vy4audn5474Q",
    "outputId": "3642e141-143b-4d25-f8dd-8cfbcf101e04"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading package lists... Done\n",
      "Building dependency tree... Done\n",
      "Reading state information... Done\n",
      "The following additional packages will be installed:\n",
      "  libhts3 libhtscodecs2\n",
      "Suggested packages:\n",
      "  python3-numpy python3-matplotlib texlive-latex-recommended\n",
      "The following NEW packages will be installed:\n",
      "  bcftools libhts3 libhtscodecs2\n",
      "0 upgraded, 3 newly installed, 0 to remove and 49 not upgraded.\n",
      "Need to get 1,140 kB of archives.\n",
      "After this operation, 3,448 kB of additional disk space will be used.\n",
      "Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhtscodecs2 amd64 1.1.1-3 [53.2 kB]\n",
      "Get:2 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhts3 amd64 1.13+ds-2build1 [390 kB]\n",
      "Get:3 http://archive.ubuntu.com/ubuntu jammy/universe amd64 bcftools amd64 1.13-1 [697 kB]\n",
      "Fetched 1,140 kB in 1s (1,223 kB/s)\n",
      "Selecting previously unselected package libhtscodecs2:amd64.\n",
      "(Reading database ... 123632 files and directories currently installed.)\n",
      "Preparing to unpack .../libhtscodecs2_1.1.1-3_amd64.deb ...\n",
      "Unpacking libhtscodecs2:amd64 (1.1.1-3) ...\n",
      "Selecting previously unselected package libhts3:amd64.\n",
      "Preparing to unpack .../libhts3_1.13+ds-2build1_amd64.deb ...\n",
      "Unpacking libhts3:amd64 (1.13+ds-2build1) ...\n",
      "Selecting previously unselected package bcftools.\n",
      "Preparing to unpack .../bcftools_1.13-1_amd64.deb ...\n",
      "Unpacking bcftools (1.13-1) ...\n",
      "Setting up libhtscodecs2:amd64 (1.1.1-3) ...\n",
      "Setting up libhts3:amd64 (1.13+ds-2build1) ...\n",
      "Setting up bcftools (1.13-1) ...\n",
      "Processing triggers for libc-bin (2.35-0ubuntu3.4) ...\n",
      "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n",
      "\n",
      "/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n",
      "\n",
      "/sbin/ldconfig.real: /usr/local/lib/libtcm.so.1 is not a symbolic link\n",
      "\n",
      "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_opencl.so.0 is not a symbolic link\n",
      "\n",
      "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero.so.0 is not a symbolic link\n",
      "\n",
      "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n",
      "\n",
      "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n",
      "\n",
      "/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n",
      "\n",
      "/sbin/ldconfig.real: /usr/local/lib/libhwloc.so.15 is not a symbolic link\n",
      "\n",
      "/sbin/ldconfig.real: /usr/local/lib/libtcm_debug.so.1 is not a symbolic link\n",
      "\n",
      "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n",
      "\n",
      "/sbin/ldconfig.real: /usr/local/lib/libur_loader.so.0 is not a symbolic link\n",
      "\n",
      "/sbin/ldconfig.real: /usr/local/lib/libumf.so.0 is not a symbolic link\n",
      "\n",
      "Processing triggers for man-db (2.10.2-1) ...\n",
      "Reading package lists... Done\n",
      "Building dependency tree... Done\n",
      "Reading state information... Done\n",
      "Suggested packages:\n",
      "  samtools\n",
      "The following NEW packages will be installed:\n",
      "  bwa\n",
      "0 upgraded, 1 newly installed, 0 to remove and 49 not upgraded.\n",
      "Need to get 195 kB of archives.\n",
      "After this operation, 466 kB of additional disk space will be used.\n",
      "Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 bwa amd64 0.7.17-6 [195 kB]\n",
      "Fetched 195 kB in 1s (333 kB/s)\n",
      "Selecting previously unselected package bwa.\n",
      "(Reading database ... 123706 files and directories currently installed.)\n",
      "Preparing to unpack .../bwa_0.7.17-6_amd64.deb ...\n",
      "Unpacking bwa (0.7.17-6) ...\n",
      "Setting up bwa (0.7.17-6) ...\n",
      "Processing triggers for man-db (2.10.2-1) ...\n",
      "Reading package lists... Done\n",
      "Building dependency tree... Done\n",
      "Reading state information... Done\n",
      "Suggested packages:\n",
      "  cwltool\n",
      "The following NEW packages will be installed:\n",
      "  samtools\n",
      "0 upgraded, 1 newly installed, 0 to remove and 49 not upgraded.\n",
      "Need to get 520 kB of archives.\n",
      "After this operation, 1,183 kB of additional disk space will be used.\n",
      "Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 samtools amd64 1.13-4 [520 kB]\n",
      "Fetched 520 kB in 1s (714 kB/s)\n",
      "Selecting previously unselected package samtools.\n",
      "(Reading database ... 123723 files and directories currently installed.)\n",
      "Preparing to unpack .../samtools_1.13-4_amd64.deb ...\n",
      "Unpacking samtools (1.13-4) ...\n",
      "Setting up samtools (1.13-4) ...\n",
      "Processing triggers for man-db (2.10.2-1) ...\n"
     ]
    }
   ],
   "source": [
    "#install necessary tools\n",
    "!apt install bcftools\n",
    "!apt install bwa\n",
    "!apt install samtools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "ETdPCH7E5xm3",
    "outputId": "9eafbb4b-6385-4d66-f913-01b93cf42475"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[bwa_index] Pack FASTA... 0.00 sec\n",
      "[bwa_index] Construct BWT for the packed sequence...\n",
      "[bwa_index] 0.01 seconds elapse.\n",
      "[bwa_index] Update BWT... 0.00 sec\n",
      "[bwa_index] Pack forward-only FASTA... 0.00 sec\n",
      "[bwa_index] Construct SA from BWT and Occ... 0.00 sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa index -p Wuhan WuhanRef.fasta\n",
      "[main] Real time: 0.060 sec; CPU: 0.016 sec\n"
     ]
    }
   ],
   "source": [
    "#use bwa to index reference gneome\n",
    "!bwa index -p Wuhan WuhanRef.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "0Dz41QxA8WB5",
    "outputId": "f300a5d5-c3ab-4d4c-e520-68afd75d5e47"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[M::process] read 131580 sequences (10000080 bp)...\n",
      "[M::process] read 131580 sequences (10000080 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (83, 63701, 238, 79)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (211, 487, 4761)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 13861)\n",
      "[M::mem_pestat] mean and std.dev: (2426.69, 2926.00)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 18411)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (173, 242, 304)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 566)\n",
      "[M::mem_pestat] mean and std.dev: (239.65, 94.08)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 697)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (668, 821, 2664)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 6656)\n",
      "[M::mem_pestat] mean and std.dev: (1343.43, 1414.72)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 8652)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (204, 701, 4074)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 11814)\n",
      "[M::mem_pestat] mean and std.dev: (2459.25, 2931.48)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 15684)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 131580 reads in 3.414 CPU sec, 3.198 real sec\n",
      "[M::process] read 131580 sequences (10000080 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (88, 63723, 200, 92)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (187, 337, 2859)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 8203)\n",
      "[M::mem_pestat] mean and std.dev: (1391.37, 2015.97)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 10875)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (172, 240, 302)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 562)\n",
      "[M::mem_pestat] mean and std.dev: (238.02, 92.46)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 692)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (689, 852, 3491)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 9095)\n",
      "[M::mem_pestat] mean and std.dev: (2109.46, 2348.11)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 11897)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (215, 865, 4334)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 12572)\n",
      "[M::mem_pestat] mean and std.dev: (2650.93, 3183.27)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 16691)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 131580 reads in 3.886 CPU sec, 3.608 real sec\n",
      "[M::process] read 131580 sequences (10000080 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (93, 63719, 218, 93)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (168, 255, 1758)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4938)\n",
      "[M::mem_pestat] mean and std.dev: (625.30, 947.84)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6528)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (173, 242, 304)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 566)\n",
      "[M::mem_pestat] mean and std.dev: (239.91, 93.64)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 697)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (684, 872, 4650)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 12582)\n",
      "[M::mem_pestat] mean and std.dev: (2618.30, 3037.78)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 16548)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (187, 291, 2914)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 8368)\n",
      "[M::mem_pestat] mean and std.dev: (1645.26, 2458.46)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 11479)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 131580 reads in 4.840 CPU sec, 5.087 real sec\n",
      "[M::process] read 131580 sequences (10000080 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (93, 63639, 240, 106)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (187, 306, 3915)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 11371)\n",
      "[M::mem_pestat] mean and std.dev: (2244.03, 2962.10)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 15099)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (174, 242, 305)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 567)\n",
      "[M::mem_pestat] mean and std.dev: (240.62, 93.89)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 698)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (678, 813, 2021)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4707)\n",
      "[M::mem_pestat] mean and std.dev: (996.68, 863.40)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6050)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (194, 515, 4701)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 13715)\n",
      "[M::mem_pestat] mean and std.dev: (2506.54, 3040.35)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 18222)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 131580 reads in 3.461 CPU sec, 3.473 real sec\n",
      "[M::process] read 131580 sequences (10000080 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (89, 63702, 211, 82)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (193, 300, 1297)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 3505)\n",
      "[M::mem_pestat] mean and std.dev: (446.99, 646.57)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 4609)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (174, 242, 305)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 567)\n",
      "[M::mem_pestat] mean and std.dev: (240.44, 94.22)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 698)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (704, 843, 4012)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 10628)\n",
      "[M::mem_pestat] mean and std.dev: (2473.13, 2792.12)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 13936)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (164, 316, 5307)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 15593)\n",
      "[M::mem_pestat] mean and std.dev: (2614.33, 3285.96)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 20736)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 131580 reads in 3.468 CPU sec, 3.195 real sec\n",
      "[M::process] read 131580 sequences (10000080 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (77, 63775, 192, 79)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (187, 342, 3035)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 8731)\n",
      "[M::mem_pestat] mean and std.dev: (1567.33, 2072.30)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 11579)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (172, 239, 301)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 559)\n",
      "[M::mem_pestat] mean and std.dev: (237.61, 92.06)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 688)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (708, 847, 4013)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 10623)\n",
      "[M::mem_pestat] mean and std.dev: (2508.21, 2800.36)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 13928)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (170, 373, 3624)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 10532)\n",
      "[M::mem_pestat] mean and std.dev: (2179.99, 2928.78)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 13986)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 131580 reads in 3.495 CPU sec, 3.195 real sec\n",
      "[M::process] read 131580 sequences (10000080 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (98, 63698, 202, 86)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (187, 306, 3043)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 8755)\n",
      "[M::mem_pestat] mean and std.dev: (1662.39, 2540.47)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 11824)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (175, 243, 306)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 568)\n",
      "[M::mem_pestat] mean and std.dev: (240.93, 94.09)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 699)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (666, 831, 3060)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 7848)\n",
      "[M::mem_pestat] mean and std.dev: (1663.14, 1935.73)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 10242)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (189, 371, 3544)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 10254)\n",
      "[M::mem_pestat] mean and std.dev: (2107.21, 2849.27)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 13609)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 131580 reads in 4.441 CPU sec, 4.242 real sec\n",
      "[M::process] read 131580 sequences (10000080 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (98, 63712, 211, 89)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (163, 253, 685)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1729)\n",
      "[M::mem_pestat] mean and std.dev: (261.42, 215.68)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2251)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (175, 243, 305)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 565)\n",
      "[M::mem_pestat] mean and std.dev: (241.46, 93.43)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 695)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (692, 852, 4340)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 11636)\n",
      "[M::mem_pestat] mean and std.dev: (2604.02, 2847.29)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 15284)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (244, 364, 4493)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 12991)\n",
      "[M::mem_pestat] mean and std.dev: (2115.29, 2739.47)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 17240)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 131580 reads in 5.001 CPU sec, 5.098 real sec\n",
      "[M::process] read 131580 sequences (10000080 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (85, 63690, 201, 114)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (199, 300, 2669)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 7609)\n",
      "[M::mem_pestat] mean and std.dev: (1448.13, 2077.95)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 10079)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (172, 240, 302)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 562)\n",
      "[M::mem_pestat] mean and std.dev: (238.63, 92.01)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 692)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (665, 809, 2855)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 7235)\n",
      "[M::mem_pestat] mean and std.dev: (1440.14, 1609.14)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 9425)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (183, 283, 4742)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 13860)\n",
      "[M::mem_pestat] mean and std.dev: (2310.11, 3035.55)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 18419)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 131580 reads in 3.854 CPU sec, 3.597 real sec\n",
      "[M::process] read 43310 sequences (3291560 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (81, 63710, 211, 110)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (198, 334, 3087)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 8865)\n",
      "[M::mem_pestat] mean and std.dev: (1572.38, 2299.19)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 11754)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (175, 243, 305)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 565)\n",
      "[M::mem_pestat] mean and std.dev: (241.02, 93.31)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 695)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (688, 866, 3562)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 9310)\n",
      "[M::mem_pestat] mean and std.dev: (2229.55, 2492.06)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 12198)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (200, 362, 3148)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 9044)\n",
      "[M::mem_pestat] mean and std.dev: (1504.22, 2200.37)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 11992)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 131580 reads in 3.970 CPU sec, 3.772 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (28, 20932, 75, 23)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (253, 447, 4431)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 12787)\n",
      "[M::mem_pestat] mean and std.dev: (2441.21, 3197.32)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 16965)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (175, 243, 305)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 565)\n",
      "[M::mem_pestat] mean and std.dev: (241.23, 92.95)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 695)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (662, 774, 949)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (88, 1523)\n",
      "[M::mem_pestat] mean and std.dev: (692.21, 190.78)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1810)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (285, 397, 3278)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 9264)\n",
      "[M::mem_pestat] mean and std.dev: (2134.52, 2647.93)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 12726)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 43310 reads in 1.537 CPU sec, 1.466 real sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem Wuhan SARS.R1.fastq SARS.R2.fastq\n",
      "[main] Real time: 40.241 sec; CPU: 41.559 sec\n"
     ]
    }
   ],
   "source": [
    "#use bwa to align wuhan and sample genomes and into a sam file\n",
    "!bwa mem Wuhan SARS.R1.fastq SARS.R2.fastq > SARS1.sam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "OQvsDNqZ9Zlp",
    "outputId": "0f4d3f12-81d7-4e21-c382-c160254daf32"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[M::process] read 39842 sequences (10000342 bp)...\n",
      "[M::process] read 39842 sequences (10000342 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (2, 19855, 0, 1)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (379, 385, 403)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (331, 451)\n",
      "[M::mem_pestat] mean and std.dev: (388.00, 10.96)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (307, 475)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 39842 reads in 2.957 CPU sec, 2.873 real sec\n",
      "[M::process] read 39842 sequences (10000342 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (4, 19848, 0, 5)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (393, 395, 408)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (363, 438)\n",
      "[M::mem_pestat] mean and std.dev: (397.05, 7.89)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (348, 453)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 39842 reads in 4.025 CPU sec, 4.042 real sec\n",
      "[M::process] read 39842 sequences (10000342 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (7, 19830, 6, 7)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (381, 392, 398)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (347, 432)\n",
      "[M::mem_pestat] mean and std.dev: (389.85, 9.99)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (330, 449)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 39842 reads in 4.073 CPU sec, 4.197 real sec\n",
      "[M::process] read 39842 sequences (10000342 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (7, 19850, 6, 15)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (379, 386, 393)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (351, 421)\n",
      "[M::mem_pestat] mean and std.dev: (386.45, 8.88)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (337, 435)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (2100, 4321, 5822)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 13266)\n",
      "[M::mem_pestat] mean and std.dev: (4025.40, 2283.88)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 16988)\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 39842 reads in 2.981 CPU sec, 2.999 real sec\n",
      "[M::process] read 39842 sequences (10000342 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (10, 19841, 9, 9)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1830, 4252, 6305)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 15255)\n",
      "[M::mem_pestat] mean and std.dev: (4031.20, 2522.20)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 19730)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (380, 385, 387)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (366, 401)\n",
      "[M::mem_pestat] mean and std.dev: (384.46, 5.76)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (359, 408)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_process_seqs] Processed 39842 reads in 2.816 CPU sec, 2.670 real sec\n",
      "[M::process] read 39842 sequences (10000342 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (13, 19842, 5, 9)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1522, 4205, 7218)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 18610)\n",
      "[M::mem_pestat] mean and std.dev: (4575.46, 2890.40)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 24306)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (382, 386, 394)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (358, 418)\n",
      "[M::mem_pestat] mean and std.dev: (390.12, 7.88)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (346, 430)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_process_seqs] Processed 39842 reads in 2.716 CPU sec, 2.561 real sec\n",
      "[M::process] read 39842 sequences (10000342 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (20, 19771, 15, 12)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (2034, 5943, 8113)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 20271)\n",
      "[M::mem_pestat] mean and std.dev: (5115.20, 3159.02)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 26350)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (381, 384, 392)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (359, 414)\n",
      "[M::mem_pestat] mean and std.dev: (385.56, 5.40)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (348, 425)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (793, 2006, 5314)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 14356)\n",
      "[M::mem_pestat] mean and std.dev: (2720.60, 2274.55)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 18877)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (4113, 5708, 8951)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 18627)\n",
      "[M::mem_pestat] mean and std.dev: (5625.25, 3063.82)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 23465)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 39842 reads in 3.329 CPU sec, 3.191 real sec\n",
      "[M::process] read 39842 sequences (10000342 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (7, 19830, 12, 13)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (377, 381, 392)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (347, 422)\n",
      "[M::mem_pestat] mean and std.dev: (383.05, 9.13)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (332, 437)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (830, 1774, 5905)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 16055)\n",
      "[M::mem_pestat] mean and std.dev: (3204.58, 3190.97)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 21130)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (910, 1489, 4418)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 11434)\n",
      "[M::mem_pestat] mean and std.dev: (3079.15, 2752.63)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 14942)\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 39842 reads in 4.038 CPU sec, 4.199 real sec\n",
      "[M::process] read 39842 sequences (10000342 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (11, 19811, 14, 17)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1238, 3866, 8967)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 24425)\n",
      "[M::mem_pestat] mean and std.dev: (4213.00, 3468.97)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 32154)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (380, 381, 388)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (364, 404)\n",
      "[M::mem_pestat] mean and std.dev: (382.98, 4.84)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (356, 412)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1412, 5008, 5563)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 13865)\n",
      "[M::mem_pestat] mean and std.dev: (3970.43, 2646.32)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 18016)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1495, 4797, 7750)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 20260)\n",
      "[M::mem_pestat] mean and std.dev: (4494.24, 3007.58)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 26515)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 39842 reads in 4.028 CPU sec, 4.276 real sec\n",
      "[M::process] read 39842 sequences (10000342 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (8, 19817, 7, 14)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (381, 391, 404)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (335, 450)\n",
      "[M::mem_pestat] mean and std.dev: (393.30, 12.75)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (312, 473)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (3953, 4811, 7588)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 14858)\n",
      "[M::mem_pestat] mean and std.dev: (5291.14, 2658.35)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 18493)\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 39842 reads in 2.629 CPU sec, 2.659 real sec\n",
      "[M::process] read 39842 sequences (10000342 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (17, 19779, 11, 17)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (3777, 6655, 7923)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 16215)\n",
      "[M::mem_pestat] mean and std.dev: (5717.76, 2541.91)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 20361)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (381, 386, 410)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (323, 468)\n",
      "[M::mem_pestat] mean and std.dev: (391.24, 14.54)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (294, 497)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (3691, 6341, 7512)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 15154)\n",
      "[M::mem_pestat] mean and std.dev: (5625.73, 2534.53)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 18975)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (3086, 5015, 6216)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 12476)\n",
      "[M::mem_pestat] mean and std.dev: (4632.18, 2531.80)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 15606)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 39842 reads in 2.819 CPU sec, 2.665 real sec\n",
      "[M::process] read 39842 sequences (10000342 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (16, 19768, 22, 9)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (4695, 6868, 7467)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 13011)\n",
      "[M::mem_pestat] mean and std.dev: (5667.00, 2561.55)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 15913)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (379, 384, 390)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (357, 412)\n",
      "[M::mem_pestat] mean and std.dev: (384.74, 8.43)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (346, 423)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1825, 4867, 7916)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 20098)\n",
      "[M::mem_pestat] mean and std.dev: (4853.73, 3211.68)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 26189)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 39842 reads in 2.782 CPU sec, 2.623 real sec\n",
      "[M::process] read 39842 sequences (10000342 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (4, 19787, 7, 9)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (379, 397, 399)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (339, 439)\n",
      "[M::mem_pestat] mean and std.dev: (392.08, 10.99)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (319, 459)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 39842 reads in 3.392 CPU sec, 3.243 real sec\n",
      "[M::process] read 39842 sequences (10000342 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (9, 19782, 8, 10)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (390, 392, 393)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (384, 399)\n",
      "[M::mem_pestat] mean and std.dev: (391.98, 2.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (381, 402)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1994, 4795, 8528)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 21596)\n",
      "[M::mem_pestat] mean and std.dev: (5444.90, 3111.86)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 28130)\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 39842 reads in 5.229 CPU sec, 5.568 real sec\n",
      "[M::process] read 39842 sequences (10000342 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (10, 19743, 14, 10)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1824, 6676, 8267)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 21153)\n",
      "[M::mem_pestat] mean and std.dev: (4993.80, 3388.46)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 27596)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (379, 391, 391)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (355, 415)\n",
      "[M::mem_pestat] mean and std.dev: (388.16, 8.36)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (343, 427)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1774, 3773, 7875)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 20077)\n",
      "[M::mem_pestat] mean and std.dev: (4362.50, 2932.67)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 26178)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1369, 6023, 7924)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 21034)\n",
      "[M::mem_pestat] mean and std.dev: (4806.50, 3081.98)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 27589)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 39842 reads in 3.988 CPU sec, 4.070 real sec\n",
      "[M::process] read 17442 sequences (4377942 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (16, 19697, 18, 16)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1836, 3069, 6718)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 16482)\n",
      "[M::mem_pestat] mean and std.dev: (3936.69, 2900.88)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 21364)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (382, 385, 390)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (366, 406)\n",
      "[M::mem_pestat] mean and std.dev: (385.16, 3.89)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (358, 414)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (814, 1771, 2409)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5599)\n",
      "[M::mem_pestat] mean and std.dev: (1737.22, 1035.46)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7194)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1237, 3706, 7422)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 19792)\n",
      "[M::mem_pestat] mean and std.dev: (4129.81, 3146.41)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 25977)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 39842 reads in 3.128 CPU sec, 3.042 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (5, 7598, 3, 4)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (379, 392, 404)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (329, 454)\n",
      "[M::mem_pestat] mean and std.dev: (391.68, 10.78)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (304, 479)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 17442 reads in 1.431 CPU sec, 1.334 real sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem Wuhan SARS-LONG-EDT.R1.fastq SARS-LONG-EDT.R2.fastq\n",
      "[main] Real time: 56.701 sec; CPU: 56.493 sec\n"
     ]
    }
   ],
   "source": [
    "!bwa mem Wuhan SARS-LONG-EDT.R1.fastq SARS-LONG-EDT.R2.fastq > SARS2.sam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "xAmvnwe1_vi2"
   },
   "outputs": [],
   "source": [
    "#use sam tools to convert same file into sorted file\n",
    "!samtools view -bS -q 12 SARS1.sam | samtools sort -o SARS1.sam.sorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "RS_AdLF2_0rN"
   },
   "outputs": [],
   "source": [
    "!samtools view -bS -q 12 SARS2.sam | samtools sort -o SARS2.sam.sorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "Pl8NmzP1AC01",
    "outputId": "75df9bdd-1d3c-4241-d4de-8df6f5bce8b6"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[mpileup] 1 samples in 1 input files\n",
      "[mpileup] maximum number of reads per input file set to -d 2000\n"
     ]
    }
   ],
   "source": [
    "#sorting the variants from the reads, using the Wuhan reference genome\n",
    "!bcftools mpileup -f WuhanRef.fasta --max-depth  2000 SARS1.sam.sorted  | bcftools call --multiallelic-caller --variants-only --ploidy 1  -mv -Oz -o SARS1.Variants.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "3uLn8vijARSA",
    "outputId": "a2b470be-1a1e-4bb4-d413-b567d6cd6ea0"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[mpileup] 1 samples in 1 input files\n",
      "[mpileup] maximum number of reads per input file set to -d 2000\n"
     ]
    }
   ],
   "source": [
    "!bcftools mpileup -f WuhanRef.fasta --max-depth  2000 SARS2.sam.sorted  | bcftools call --multiallelic-caller --variants-only --ploidy 1  -mv -Oz -o SARS2.Variants.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "bDGjWHDuA578"
   },
   "outputs": [],
   "source": [
    "#gzip opens the VCF files\n",
    "!gzip -df SARS1.Variants.vcf.gz\n",
    "!gzip -df SARS2.Variants.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "fIkS-fK3A8Ni",
    "outputId": "3c045efa-4106-4c73-a2d8-9d0c8dbd33e0"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##fileformat=VCFv4.2\n",
      "##FILTER=<ID=PASS,Description=\"All filters passed\">\n",
      "##bcftoolsVersion=1.13+htslib-1.13+ds\n",
      "##bcftoolsCommand=mpileup -f WuhanRef.fasta --max-depth 2000 SARS1.sam.sorted\n",
      "##reference=file://WuhanRef.fasta\n",
      "##contig=<ID=NC_045512.2,length=29903>\n",
      "##ALT=<ID=*,Description=\"Represents allele(s) other than observed.\">\n",
      "##INFO=<ID=INDEL,Number=0,Type=Flag,Description=\"Indicates that the variant is an INDEL.\">\n",
      "##INFO=<ID=IDV,Number=1,Type=Integer,Description=\"Maximum number of raw reads supporting an indel\">\n",
      "##INFO=<ID=IMF,Number=1,Type=Float,Description=\"Maximum fraction of raw reads supporting an indel\">\n",
      "##INFO=<ID=DP,Number=1,Type=Integer,Description=\"Raw read depth\">\n",
      "##INFO=<ID=VDB,Number=1,Type=Float,Description=\"Variant Distance Bias for filtering splice-site artefacts in RNA-seq data (bigger is better)\",Version=\"3\">\n",
      "##INFO=<ID=RPBZ,Number=1,Type=Float,Description=\"Mann-Whitney U-z test of Read Position Bias (closer to 0 is better)\">\n",
      "##INFO=<ID=MQBZ,Number=1,Type=Float,Description=\"Mann-Whitney U-z test of Mapping Quality Bias (closer to 0 is better)\">\n",
      "##INFO=<ID=BQBZ,Number=1,Type=Float,Description=\"Mann-Whitney U-z test of Base Quality Bias (closer to 0 is better)\">\n",
      "##INFO=<ID=MQSBZ,Number=1,Type=Float,Description=\"Mann-Whitney U-z test of Mapping Quality vs Strand Bias (closer to 0 is better)\">\n",
      "##INFO=<ID=SCBZ,Number=1,Type=Float,Description=\"Mann-Whitney U-z test of Soft-Clip Length Bias (closer to 0 is better)\">\n",
      "##INFO=<ID=FS,Number=1,Type=Float,Description=\"Phred-scaled p-value using Fisher's exact test to detect strand bias\">\n",
      "##INFO=<ID=SGB,Number=1,Type=Float,Description=\"Segregation based metric.\">\n",
      "##INFO=<ID=MQ0F,Number=1,Type=Float,Description=\"Fraction of MQ0 reads (smaller is better)\">\n",
      "##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"List of Phred-scaled genotype likelihoods\">\n",
      "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
      "##INFO=<ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "##INFO=<ID=AN,Number=1,Type=Integer,Description=\"Total number of alleles in called genotypes\">\n",
      "##INFO=<ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "##INFO=<ID=MQ,Number=1,Type=Integer,Description=\"Average mapping quality\">\n",
      "##bcftools_callVersion=1.13+htslib-1.13+ds\n",
      "##bcftools_callCommand=call --multiallelic-caller --variants-only --ploidy 1 -mv -Oz -o SARS1.Variants.vcf.gz; Date=Fri Dec  6 03:25:53 2024\n",
      "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tSARS1.sam.sorted\n",
      "NC_045512.2\t21\t.\tC\tT\t54.4146\t.\tDP=3;VDB=0.364077;SGB=-0.511536;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,2,1;MQ=60\tGT:PL\t1:84,0\n",
      "NC_045512.2\t241\t.\tC\tT\t228.428\t.\tDP=1568;VDB=0.822957;SGB=-0.693147;RPBZ=-0.880374;MQBZ=0;MQSBZ=0;BQBZ=-0.331577;SCBZ=0.0226572;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,3,702,771;MQ=60\tGT:PL\t1:255,0\n",
      "NC_045512.2\t1387\t.\tC\tT\t225.422\t.\tDP=1934;VDB=0.00106222;SGB=-0.693147;MQSBZ=0.0295856;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,1343,498;MQ=59\tGT:PL\t1:255,0\n",
      "NC_045512.2\t2832\t.\tA\tG\t228.426\t.\tDP=1192;VDB=0.999976;SGB=-0.693147;RPBZ=0.198349;BQBZ=-0.553437;SCBZ=0.500841;FS=0;MQ0F=0;AC=1;AN=1;DP4=3,0,395,755;MQ=60\tGT:PL\t1:255,0\n",
      "NC_045512.2\t3037\t.\tC\tT\t225.417\t.\tDP=215;VDB=0.999998;SGB=-0.693147;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,81,107;MQ=60\tGT:PL\t1:255,0\n",
      "NC_045512.2\t5386\t.\tT\tG\t228.426\t.\tDP=697;VDB=0.00583288;SGB=-0.693147;RPBZ=0.397928;MQBZ=-0.0539949;MQSBZ=0.823437;BQBZ=-0.320427;SCBZ=0.121088;FS=0;MQ0F=0;AC=1;AN=1;DP4=1,1,409,277;MQ=59\tGT:PL\t1:255,0\n",
      "NC_045512.2\t8393\t.\tG\tA\t225.417\t.\tDP=240;VDB=0.900753;SGB=-0.693147;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,204,33;MQ=60\tGT:PL\t1:255,0\n",
      "NC_045512.2\t10029\t.\tC\tT\t225.426\t.\tDP=1970;VDB=0.981163;SGB=-0.693147;MQSBZ=-0.0271871;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,703,1197;MQ=59\tGT:PL\t1:255,0\n",
      "NC_045512.2\t10449\t.\tC\tA\t201.427\t.\tDP=1508;VDB=1;SGB=-0.693147;RPBZ=1.38851;MQBZ=0;MQSBZ=0;BQBZ=-1.67342;SCBZ=0.0192999;FS=0;MQ0F=0;AC=1;AN=1;DP4=3,0,1455,34;MQ=60\tGT:PL\t1:228,0\n",
      "NC_045512.2\t11291\t.\tG\tA\t35.1339\t.\tDP=303;VDB=0;SGB=-0.693147;RPBZ=-6.49767;BQBZ=-10.1122;SCBZ=-10.6298;FS=0;MQ0F=0;AC=1;AN=1;DP4=2,6,11,275;MQ=60\tGT:PL\t1:62,0\n",
      "NC_045512.2\t11537\t.\tA\tG\t225.426\t.\tDP=1973;VDB=0.47178;SGB=-0.693147;MQSBZ=-0.0315124;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,836,1100;MQ=59\tGT:PL\t1:255,0\n",
      "NC_045512.2\t13195\t.\tT\tC\t228.435\t.\tDP=1107;VDB=0.000372741;SGB=-0.693147;RPBZ=-1.62498;BQBZ=0.0120568;SCBZ=0.0809581;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,1,650,424;MQ=60\tGT:PL\t1:255,0\n",
      "NC_045512.2\t14408\t.\tC\tT\t225.417\t.\tDP=373;VDB=0.375702;SGB=-0.693147;MQSBZ=0.749429;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,219,123;MQ=59\tGT:PL\t1:255,0\n",
      "NC_045512.2\t15240\t.\tC\tT\t228.429\t.\tDP=1993;VDB=4.19798e-33;SGB=-0.693147;RPBZ=-2.28037;MQBZ=0;MQSBZ=0;BQBZ=0.0639893;SCBZ=0.0706901;FS=0;MQ0F=0;AC=1;AN=1;DP4=4,3,1888,57;MQ=60\tGT:PL\t1:255,0\n",
      "NC_045512.2\t18163\t.\tA\tG\t225.417\t.\tDP=782;VDB=5.91011e-07;SGB=-0.693147;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,401,354;MQ=60\tGT:PL\t1:255,0\n",
      "NC_045512.2\t21762\t.\tC\tT\t89.419\t.\tDP=1169;VDB=1.40258e-15;SGB=-0.693147;RPBZ=1.01145;MQBZ=-0.0603022;MQSBZ=-0.286778;BQBZ=-5.80241;SCBZ=0.0854351;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,1,630,473;MQ=59\tGT:PL\t1:116,0\n",
      "NC_045512.2\t21846\t.\tC\tT\t225.426\t.\tDP=1639;VDB=0.78749;SGB=-0.693147;MQSBZ=0;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,940,664;MQ=60\tGT:PL\t1:255,0\n",
      "NC_045512.2\t22578\t.\tG\tA\t228.424\t.\tDP=659;VDB=4.58797e-08;SGB=-0.693147;RPBZ=1.73192;BQBZ=0.304704;SCBZ=-6.04041;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,1,477,127;MQ=59\tGT:PL\t1:255,0\n",
      "NC_045512.2\t22673\t.\tT\tC\t228.434\t.\tDP=747;VDB=0.00094316;SGB=-0.693147;RPBZ=-1.13685;MQBZ=-0.0372678;MQSBZ=0.358225;BQBZ=0.161276;SCBZ=0.130091;FS=0;MQ0F=0;AC=1;AN=1;DP4=1,0,638,82;MQ=59\tGT:PL\t1:255,0\n",
      "NC_045512.2\t22674\t.\tC\tT\t225.417\t.\tDP=744;VDB=0.00103229;SGB=-0.693147;MQSBZ=0.358225;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,639,82;MQ=59\tGT:PL\t1:255,0\n",
      "NC_045512.2\t22679\t.\tT\tC\t225.422\t.\tDP=769;VDB=0.00304436;SGB=-0.693147;MQSBZ=0.402114;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,637,103;MQ=59\tGT:PL\t1:255,0\n",
      "NC_045512.2\t22686\t.\tC\tT\t225.422\t.\tDP=736;VDB=1.70409e-11;SGB=-0.693147;MQSBZ=0.486713;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,553,131;MQ=59\tGT:PL\t1:255,0\n",
      "NC_045512.2\t22813\t.\tG\tT\t228.409\t.\tDP=776;VDB=0.145068;SGB=-0.693147;RPBZ=3.85042;BQBZ=0.116351;SCBZ=-2.76578;FS=0;MQ0F=0;AC=1;AN=1;DP4=7,1,272,453;MQ=60\tGT:PL\t1:255,0\n",
      "NC_045512.2\t22898\t.\tG\tA\t134.818\t.\tDP=489;VDB=0.0946854;SGB=-0.693147;RPBZ=-5.72535;BQBZ=-2.54205;SCBZ=-1.64068;FS=0;MQ0F=0;AC=1;AN=1;DP4=37,87,199,162;MQ=60\tGT:PL\t1:222,60\n",
      "NC_045512.2\t22992\t.\tG\tA\t225.422\t.\tDP=300;VDB=0.9999;SGB=-0.693147;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,119,174;MQ=60\tGT:PL\t1:255,0\n",
      "NC_045512.2\t22995\t.\tC\tA\t225.417\t.\tDP=291;VDB=0.999496;SGB=-0.693147;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,99,185;MQ=60\tGT:PL\t1:255,0\n",
      "NC_045512.2\t23013\t.\tA\tC\t225.426\t.\tDP=361;VDB=0.684437;SGB=-0.693147;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,107,245;MQ=60\tGT:PL\t1:255,0\n",
      "NC_045512.2\t23040\t.\tA\tG\t225.422\t.\tDP=411;VDB=4.34195e-06;SGB=-0.693147;MQSBZ=-0.571717;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,101,309;MQ=59\tGT:PL\t1:255,0\n",
      "NC_045512.2\t23048\t.\tG\tA\t225.426\t.\tDP=438;VDB=0.00108106;SGB=-0.693147;MQSBZ=-0.624321;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,122,313;MQ=59\tGT:PL\t1:255,0\n",
      "NC_045512.2\t23055\t.\tA\tG\t228.427\t.\tDP=488;VDB=0.264075;SGB=-0.693147;RPBZ=-0.724204;MQBZ=-0.0455961;MQSBZ=-0.559778;BQBZ=-0.205293;SCBZ=0.228902;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,1,115,366;MQ=59\tGT:PL\t1:255,0\n",
      "NC_045512.2\t23063\t.\tA\tT\t225.422\t.\tDP=514;VDB=0.992994;SGB=-0.693147;MQSBZ=-0.648204;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,150,357;MQ=59\tGT:PL\t1:255,0\n",
      "NC_045512.2\t23075\t.\tT\tC\t228.427\t.\tDP=478;VDB=0.973092;SGB=-0.693147;RPBZ=0.569017;MQBZ=-0.0463241;MQSBZ=-0.624404;BQBZ=-0.0458501;SCBZ=0.123354;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,1,131,335;MQ=59\tGT:PL\t1:255,0\n",
      "NC_045512.2\t23202\t.\tC\tA\t228.419\t.\tDP=1980;VDB=1.80227e-27;SGB=-0.693147;RPBZ=-2.27746;MQBZ=0;MQSBZ=0;BQBZ=-0.15552;SCBZ=-0.417443;FS=0;MQ0F=0;AC=1;AN=1;DP4=3,9,1810,108;MQ=60\tGT:PL\t1:255,0\n",
      "NC_045512.2\t23403\t.\tA\tG\t228.436\t.\tDP=1998;VDB=7.37984e-11;SGB=-0.693147;RPBZ=-0.0136828;MQBZ=-0.000814724;MQSBZ=-0.0198889;BQBZ=0.0890505;SCBZ=0.0302742;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,1,454,1448;MQ=59\tGT:PL\t1:255,0\n",
      "NC_045512.2\t23525\t.\tC\tT\t225.417\t.\tDP=406;VDB=1;SGB=-0.693147;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,363,29;MQ=60\tGT:PL\t1:255,0\n",
      "NC_045512.2\t23599\t.\tT\tG\t225.417\t.\tDP=202;VDB=0.314335;SGB=-0.693147;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,143,48;MQ=60\tGT:PL\t1:255,0\n",
      "NC_045512.2\t23604\t.\tC\tA\t225.422\t.\tDP=194;VDB=0.10292;SGB=-0.693147;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,137,46;MQ=60\tGT:PL\t1:255,0\n",
      "NC_045512.2\t23854\t.\tC\tA\t228.431\t.\tDP=1970;VDB=0;SGB=-0.693147;RPBZ=0.0148588;MQBZ=0;MQSBZ=0;BQBZ=0.047141;SCBZ=0.0343282;FS=0;MQ0F=0;AC=1;AN=1;DP4=1,0,1843,45;MQ=59\tGT:PL\t1:255,0\n",
      "NC_045512.2\t23948\t.\tG\tT\t225.422\t.\tDP=1989;VDB=0.95685;SGB=-0.693147;MQSBZ=0.0256857;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,1241,654;MQ=59\tGT:PL\t1:255,0\n",
      "NC_045512.2\t24130\t.\tC\tA\t225.422\t.\tDP=1624;VDB=6.18528e-05;SGB=-0.693147;MQSBZ=0;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,321,1238;MQ=60\tGT:PL\t1:255,0\n",
      "NC_045512.2\t24424\t.\tA\tT\t228.425\t.\tDP=1984;VDB=5.75948e-16;SGB=-0.693147;RPBZ=0.558542;MQBZ=0;MQSBZ=0;BQBZ=0.00458417;SCBZ=-2.1878;FS=0;MQ0F=0;AC=1;AN=1;DP4=2,0,1576,386;MQ=59\tGT:PL\t1:255,0\n",
      "NC_045512.2\t24469\t.\tT\tA\t225.417\t.\tDP=1992;VDB=9.3932e-05;SGB=-0.693147;MQSBZ=0;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,1915,43;MQ=60\tGT:PL\t1:255,0\n",
      "NC_045512.2\t24503\t.\tC\tT\t228.427\t.\tDP=1746;VDB=0.00246117;SGB=-0.693147;RPBZ=-0.674078;MQBZ=0;MQSBZ=0;BQBZ=0.889907;SCBZ=0.0696336;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,4,1305,385;MQ=60\tGT:PL\t1:255,0\n",
      "NC_045512.2\t25584\t.\tC\tT\t225.417\t.\tDP=391;VDB=4.66957e-09;SGB=-0.693147;MQSBZ=-0.332363;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,38,344;MQ=59\tGT:PL\t1:255,0\n",
      "NC_045512.2\t26270\t.\tC\tT\t225.426\t.\tDP=1992;VDB=2.33557e-07;SGB=-0.693147;MQSBZ=0.0113088;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,1802,170;MQ=59\tGT:PL\t1:255,0\n",
      "NC_045512.2\t26530\t.\tA\tG\t228.376\t.\tDP=1994;VDB=0.000172602;SGB=-0.693147;RPBZ=7.78173;MQBZ=0;MQSBZ=0;BQBZ=-0.192841;SCBZ=0.29102;FS=0;MQ0F=0;AC=1;AN=1;DP4=42,13,715,1211;MQ=60\tGT:PL\t1:255,0\n",
      "NC_045512.2\t26709\t.\tG\tA\t225.426\t.\tDP=1996;VDB=2.22432e-12;SGB=-0.693147;MQSBZ=0;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,734,1230;MQ=60\tGT:PL\t1:255,0\n",
      "NC_045512.2\t27259\t.\tA\tC\t228.417\t.\tDP=309;VDB=0.813993;SGB=-0.693147;RPBZ=0.465641;BQBZ=-0.299598;SCBZ=-4.1332;FS=0;MQ0F=0;AC=1;AN=1;DP4=1,1,276,26;MQ=60\tGT:PL\t1:255,0\n",
      "NC_045512.2\t27807\t.\tC\tT\t228.416\t.\tDP=1204;VDB=7.40739e-11;SGB=-0.693147;RPBZ=3.16665;BQBZ=0.310214;SCBZ=0.589861;FS=0;MQ0F=0;AC=1;AN=1;DP4=5,1,106,1041;MQ=59\tGT:PL\t1:255,0\n",
      "NC_045512.2\t28271\t.\tA\tT\t228.433\t.\tDP=1329;VDB=2.77754e-23;SGB=-0.693147;RPBZ=1.83098;MQBZ=-0.00135301;MQSBZ=-0.0246022;BQBZ=0.0804952;SCBZ=-2.36826;FS=0;MQ0F=0;AC=1;AN=1;DP4=1,1,439,870;MQ=59\tGT:PL\t1:255,0\n",
      "NC_045512.2\t28311\t.\tC\tT\t228.429\t.\tDP=1974;VDB=6.94544e-09;SGB=-0.693147;RPBZ=-0.386016;MQBZ=-0.00232366;MQSBZ=-0.0224533;BQBZ=1.62859;SCBZ=-1.11035;FS=0;MQ0F=0;AC=1;AN=1;DP4=3,5,537,1383;MQ=59\tGT:PL\t1:255,0\n",
      "NC_045512.2\t28363\t.\tA\tT\t10.5352\t.\tDP=17;VDB=3.00589e-11;SGB=-0.689466;RPBZ=1.94365;BQBZ=-1.94365;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,1,9,7;MQ=60\tGT:PL\t1:37,0\n",
      "NC_045512.2\t28370\t.\tA\tG\t76.405\t.\tDP=398;VDB=0;SGB=-0.693147;RPBZ=-13.6931;BQBZ=-19.3391;SCBZ=0.089802;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,1,11,363;MQ=60\tGT:PL\t1:103,0\n",
      "NC_045512.2\t28881\t.\tG\tA\t225.426\t.\tDP=1530;VDB=1.39283e-14;SGB=-0.693147;MQSBZ=0.00973001;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,650,834;MQ=59\tGT:PL\t1:255,0\n",
      "NC_045512.2\t28882\t.\tG\tA\t228.436\t.\tDP=1529;VDB=2.01584e-14;SGB=-0.693147;RPBZ=-1.3178;MQBZ=-0.00203243;MQSBZ=0.00978414;BQBZ=1.45921;SCBZ=0.0140994;FS=0;MQ0F=0;AC=1;AN=1;DP4=1,0,648,834;MQ=59\tGT:PL\t1:255,0\n",
      "NC_045512.2\t28883\t.\tG\tC\t228.434\t.\tDP=1759;VDB=1.07134e-06;SGB=-0.693147;RPBZ=-1.9443;MQBZ=-0.00377274;MQSBZ=0.022165;BQBZ=1.32882;SCBZ=0.0260981;FS=0;MQ0F=0;AC=1;AN=1;DP4=3,0,644,1056;MQ=59\tGT:PL\t1:255,0\n",
      "##fileformat=VCFv4.2\n",
      "##FILTER=<ID=PASS,Description=\"All filters passed\">\n",
      "##bcftoolsVersion=1.13+htslib-1.13+ds\n",
      "##bcftoolsCommand=mpileup -f WuhanRef.fasta --max-depth 2000 SARS2.sam.sorted\n",
      "##reference=file://WuhanRef.fasta\n",
      "##contig=<ID=NC_045512.2,length=29903>\n",
      "##ALT=<ID=*,Description=\"Represents allele(s) other than observed.\">\n",
      "##INFO=<ID=INDEL,Number=0,Type=Flag,Description=\"Indicates that the variant is an INDEL.\">\n",
      "##INFO=<ID=IDV,Number=1,Type=Integer,Description=\"Maximum number of raw reads supporting an indel\">\n",
      "##INFO=<ID=IMF,Number=1,Type=Float,Description=\"Maximum fraction of raw reads supporting an indel\">\n",
      "##INFO=<ID=DP,Number=1,Type=Integer,Description=\"Raw read depth\">\n",
      "##INFO=<ID=VDB,Number=1,Type=Float,Description=\"Variant Distance Bias for filtering splice-site artefacts in RNA-seq data (bigger is better)\",Version=\"3\">\n",
      "##INFO=<ID=RPBZ,Number=1,Type=Float,Description=\"Mann-Whitney U-z test of Read Position Bias (closer to 0 is better)\">\n",
      "##INFO=<ID=MQBZ,Number=1,Type=Float,Description=\"Mann-Whitney U-z test of Mapping Quality Bias (closer to 0 is better)\">\n",
      "##INFO=<ID=BQBZ,Number=1,Type=Float,Description=\"Mann-Whitney U-z test of Base Quality Bias (closer to 0 is better)\">\n",
      "##INFO=<ID=MQSBZ,Number=1,Type=Float,Description=\"Mann-Whitney U-z test of Mapping Quality vs Strand Bias (closer to 0 is better)\">\n",
      "##INFO=<ID=SCBZ,Number=1,Type=Float,Description=\"Mann-Whitney U-z test of Soft-Clip Length Bias (closer to 0 is better)\">\n",
      "##INFO=<ID=FS,Number=1,Type=Float,Description=\"Phred-scaled p-value using Fisher's exact test to detect strand bias\">\n",
      "##INFO=<ID=SGB,Number=1,Type=Float,Description=\"Segregation based metric.\">\n",
      "##INFO=<ID=MQ0F,Number=1,Type=Float,Description=\"Fraction of MQ0 reads (smaller is better)\">\n",
      "##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"List of Phred-scaled genotype likelihoods\">\n",
      "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
      "##INFO=<ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "##INFO=<ID=AN,Number=1,Type=Integer,Description=\"Total number of alleles in called genotypes\">\n",
      "##INFO=<ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "##INFO=<ID=MQ,Number=1,Type=Integer,Description=\"Average mapping quality\">\n",
      "##bcftools_callVersion=1.13+htslib-1.13+ds\n",
      "##bcftools_callCommand=call --multiallelic-caller --variants-only --ploidy 1 -mv -Oz -o SARS2.Variants.vcf.gz; Date=Fri Dec  6 03:26:30 2024\n",
      "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tSARS2.sam.sorted\n",
      "NC_045512.2\t241\t.\tC\tT\t228.436\t.\tDP=2054;VDB=0;SGB=-0.693147;RPBZ=-0.0386264;MQBZ=-0.00119255;MQSBZ=0.00545195;BQBZ=2.36329;SCBZ=0.0430366;FS=0;MQ0F=0;AC=1;AN=1;DP4=2,0,1995,41;MQ=59\tGT:PL\t1:255,0\n",
      "NC_045512.2\t3037\t.\tC\tT\t225.417\t.\tDP=1048;VDB=0.946775;SGB=-0.693147;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,274,251;MQ=60\tGT:PL\t1:255,0\n",
      "NC_045512.2\t11842\t.\tG\tA\t225.422\t.\tDP=1876;VDB=0;SGB=-0.693147;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,0,1875;MQ=60\tGT:PL\t1:255,0\n",
      "NC_045512.2\t12759\t.\tA\tG\t225.422\t.\tDP=2015;VDB=0;SGB=-0.693147;MQSBZ=-0.139288;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,1878,137;MQ=59\tGT:PL\t1:255,0\n",
      "NC_045512.2\t14408\t.\tC\tT\t228.427\t.\tDP=1999;VDB=0;SGB=-0.693147;RPBZ=-1.81181;MQBZ=0;MQSBZ=0;BQBZ=1.54928;SCBZ=-2.0502;FS=0;MQ0F=0;AC=1;AN=1;DP4=1,0,1711,116;MQ=60\tGT:PL\t1:255,0\n",
      "NC_045512.2\t18292\t.\tG\tT\t225.426\t.\tDP=2010;VDB=2.8727e-16;SGB=-0.693147;MQSBZ=0;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,1723,287;MQ=60\tGT:PL\t1:255,0\n",
      "NC_045512.2\t20268\t.\tA\tG\t225.426\t.\tDP=707;VDB=0;SGB=-0.693147;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,707,0;MQ=60\tGT:PL\t1:255,0\n",
      "NC_045512.2\t23403\t.\tA\tG\t228.433\t.\tDP=808;VDB=0;SGB=-0.693147;RPBZ=1.87863;MQBZ=-0.0610468;MQSBZ=-0.0352017;BQBZ=4.69175;SCBZ=-3.3091;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,3,1,804;MQ=59\tGT:PL\t1:255,0\n",
      "NC_045512.2\t24978\t.\tA\tT\t5.75677\t.\tDP=1;SGB=-0.379885;FS=0;MQ0F=0;AC=1;AN=1;DP4=0,0,1,0;MQ=60\tGT:PL\t1:34,0\n",
      "NC_045512.2\t27813\t.\tC\tT\t228.426\t.\tDP=2011;VDB=0;SGB=-0.693147;RPBZ=0.0432847;MQBZ=0;MQSBZ=0;BQBZ=2.70044;SCBZ=0.0105973;FS=0;MQ0F=0;AC=1;AN=1;DP4=2,0,1963,44;MQ=60\tGT:PL\t1:255,0\n",
      "NC_045512.2\t28854\t.\tC\tT\t228.431\t.\tDP=2016;VDB=0;SGB=-0.693147;RPBZ=-0.142112;MQBZ=0;MQSBZ=0;BQBZ=2.61426;SCBZ=0.0310324;FS=0;MQ0F=0;AC=1;AN=1;DP4=3,0,1902,47;MQ=60\tGT:PL\t1:255,0\n"
     ]
    }
   ],
   "source": [
    "!cat SARS1.Variants.vcf\n",
    "!cat SARS2.Variants.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "xO7Nug-WElL3"
   },
   "outputs": [],
   "source": [
    "#install cyvcf2 to help read vcf files\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "16VX5HsqA-1G",
    "outputId": "0d9e356c-946b-4b6c-b7b8-8095cf20ae14"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting cyvcf2\n",
      "  Downloading cyvcf2-0.31.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (5.5 kB)\n",
      "Requirement already satisfied: numpy>=1.16.0 in /usr/local/lib/python3.10/dist-packages (from cyvcf2) (1.26.4)\n",
      "Collecting coloredlogs (from cyvcf2)\n",
      "  Downloading coloredlogs-15.0.1-py2.py3-none-any.whl.metadata (12 kB)\n",
      "Requirement already satisfied: click in /usr/local/lib/python3.10/dist-packages (from cyvcf2) (8.1.7)\n",
      "Collecting humanfriendly>=9.1 (from coloredlogs->cyvcf2)\n",
      "  Downloading humanfriendly-10.0-py2.py3-none-any.whl.metadata (9.2 kB)\n",
      "Downloading cyvcf2-0.31.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (6.8 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m6.8/6.8 MB\u001b[0m \u001b[31m49.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading coloredlogs-15.0.1-py2.py3-none-any.whl (46 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m46.0/46.0 kB\u001b[0m \u001b[31m2.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading humanfriendly-10.0-py2.py3-none-any.whl (86 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m86.8/86.8 kB\u001b[0m \u001b[31m5.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hInstalling collected packages: humanfriendly, coloredlogs, cyvcf2\n",
      "Successfully installed coloredlogs-15.0.1 cyvcf2-0.31.1 humanfriendly-10.0\n"
     ]
    }
   ],
   "source": [
    "!pip install cyvcf2\n",
    "vcf_path_1 = 'SARS1.Variants.vcf'\n",
    "import cyvcf2\n",
    "vcf_reader_1 = cyvcf2.VCF(vcf_path_1)\n",
    "\n",
    "with open('output1.tsv', 'w') as file:\n",
    "    for variant in vcf_reader_1:\n",
    "        genotype = variant.gt_types[0]\n",
    "        data_string = f\"{variant.CHROM}\\t{variant.POS}\\t{variant.ID}\\t{variant.REF}\\t{','.join(variant.ALT)}\\t{variant.QUAL}\\t{variant.FILTER}\\t{genotype}\\n\"\n",
    "        file.write(data_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "hGuo0nO6BDri"
   },
   "outputs": [],
   "source": [
    "vcf_path_2 = 'SARS2.Variants.vcf'\n",
    "import cyvcf2\n",
    "vcf_reader_2 = cyvcf2.VCF(vcf_path_2)\n",
    "\n",
    "with open('output2.tsv', 'w') as file:\n",
    "    for variant in vcf_reader_2:\n",
    "        genotype = variant.gt_types[0]\n",
    "        data_string = f\"{variant.CHROM}\\t{variant.POS}\\t{variant.ID}\\t{variant.REF}\\t{','.join(variant.ALT)}\\t{variant.QUAL}\\t{variant.FILTER}\\t{genotype}\\n\"\n",
    "        file.write(data_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 1000
    },
    "collapsed": true,
    "id": "ZoCDGiv0BK1R",
    "outputId": "a5a69909-680d-464d-90bf-b4efe321c3b0"
   },
   "outputs": [
    {
     "data": {
      "application/vnd.google.colaboratory.intrinsic+json": {
       "summary": "{\n  \"name\": \"SARS1_df\",\n  \"rows\": 56,\n  \"fields\": [\n    {\n      \"column\": \"CHROM\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"NC_045512.2\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"POS\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 8078,\n        \"min\": 21,\n        \"max\": 28883,\n        \"num_unique_values\": 56,\n        \"samples\": [\n          21\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"ID\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": null,\n        \"min\": null,\n        \"max\": null,\n        \"num_unique_values\": 0,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"REF\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 4,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"ALT\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 4,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"QUAL\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 51.616742228243794,\n        \"min\": 10.535200119018556,\n        \"max\": 228.43600463867188,\n        \"num_unique_values\": 26,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"FILTER\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": null,\n        \"min\": null,\n        \"max\": null,\n        \"num_unique_values\": 0,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"GENOTYPE\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0,\n        \"min\": 3,\n        \"max\": 3,\n        \"num_unique_values\": 1,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}",
       "type": "dataframe",
       "variable_name": "SARS1_df"
      },
      "text/html": [
       "\n",
       "  <div id=\"df-9005ef8b-a8e4-4f58-ad2e-1c52c511629a\" class=\"colab-df-container\">\n",
       "    <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>GENOTYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>54.414600</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>241</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>228.427994</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>1387</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>225.421997</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>2832</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>228.425995</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>3037</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>225.417007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>5386</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>228.425995</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>8393</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>225.417007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>10029</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>225.425995</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>10449</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>201.427002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>11291</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>35.133900</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>11537</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>225.425995</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>13195</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>228.434998</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>14408</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>225.417007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>15240</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>228.429001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>18163</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>225.417007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>21762</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>89.418999</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>21846</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>225.425995</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>22578</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>228.423996</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>22673</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>228.434006</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>22674</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>225.417007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>22679</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>225.421997</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>22686</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>225.421997</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>22813</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>228.408997</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>22898</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>134.817993</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>22992</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>225.421997</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>22995</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>225.417007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>23013</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>225.425995</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>23040</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>225.421997</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>23048</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>225.425995</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>23055</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>228.427002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>23063</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>225.421997</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>23075</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>228.427002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>23202</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>228.419006</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>23403</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>228.436005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>23525</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>225.417007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>23599</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>225.417007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>23604</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>225.421997</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>23854</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>228.431000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>23948</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>225.421997</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>24130</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>225.421997</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>24424</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>228.425003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>24469</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>225.417007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>24503</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>228.427002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>25584</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>225.417007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>26270</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>225.425995</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>26530</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>228.376007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>26709</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>225.425995</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>27259</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>228.417007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>27807</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>228.416000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>28271</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>228.432999</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>28311</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>228.429001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>28363</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>10.535200</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>28370</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>76.404999</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>28881</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>225.425995</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>28882</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>228.436005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>28883</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>228.434006</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "    <div class=\"colab-df-buttons\">\n",
       "\n",
       "  <div class=\"colab-df-container\">\n",
       "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-9005ef8b-a8e4-4f58-ad2e-1c52c511629a')\"\n",
       "            title=\"Convert this dataframe to an interactive table.\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
       "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
       "  </svg>\n",
       "    </button>\n",
       "\n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    .colab-df-buttons div {\n",
       "      margin-bottom: 4px;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "    <script>\n",
       "      const buttonEl =\n",
       "        document.querySelector('#df-9005ef8b-a8e4-4f58-ad2e-1c52c511629a button.colab-df-convert');\n",
       "      buttonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "      async function convertToInteractive(key) {\n",
       "        const element = document.querySelector('#df-9005ef8b-a8e4-4f58-ad2e-1c52c511629a');\n",
       "        const dataTable =\n",
       "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                    [key], {});\n",
       "        if (!dataTable) return;\n",
       "\n",
       "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "          + ' to learn more about interactive tables.';\n",
       "        element.innerHTML = '';\n",
       "        dataTable['output_type'] = 'display_data';\n",
       "        await google.colab.output.renderOutput(dataTable, element);\n",
       "        const docLink = document.createElement('div');\n",
       "        docLink.innerHTML = docLinkHtml;\n",
       "        element.appendChild(docLink);\n",
       "      }\n",
       "    </script>\n",
       "  </div>\n",
       "\n",
       "\n",
       "<div id=\"df-bcdcdbbe-db36-4ef8-af6d-40562fa763a1\">\n",
       "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-bcdcdbbe-db36-4ef8-af6d-40562fa763a1')\"\n",
       "            title=\"Suggest charts\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "     width=\"24px\">\n",
       "    <g>\n",
       "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
       "    </g>\n",
       "</svg>\n",
       "  </button>\n",
       "\n",
       "<style>\n",
       "  .colab-df-quickchart {\n",
       "      --bg-color: #E8F0FE;\n",
       "      --fill-color: #1967D2;\n",
       "      --hover-bg-color: #E2EBFA;\n",
       "      --hover-fill-color: #174EA6;\n",
       "      --disabled-fill-color: #AAA;\n",
       "      --disabled-bg-color: #DDD;\n",
       "  }\n",
       "\n",
       "  [theme=dark] .colab-df-quickchart {\n",
       "      --bg-color: #3B4455;\n",
       "      --fill-color: #D2E3FC;\n",
       "      --hover-bg-color: #434B5C;\n",
       "      --hover-fill-color: #FFFFFF;\n",
       "      --disabled-bg-color: #3B4455;\n",
       "      --disabled-fill-color: #666;\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart {\n",
       "    background-color: var(--bg-color);\n",
       "    border: none;\n",
       "    border-radius: 50%;\n",
       "    cursor: pointer;\n",
       "    display: none;\n",
       "    fill: var(--fill-color);\n",
       "    height: 32px;\n",
       "    padding: 0;\n",
       "    width: 32px;\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart:hover {\n",
       "    background-color: var(--hover-bg-color);\n",
       "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "    fill: var(--button-hover-fill-color);\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart-complete:disabled,\n",
       "  .colab-df-quickchart-complete:disabled:hover {\n",
       "    background-color: var(--disabled-bg-color);\n",
       "    fill: var(--disabled-fill-color);\n",
       "    box-shadow: none;\n",
       "  }\n",
       "\n",
       "  .colab-df-spinner {\n",
       "    border: 2px solid var(--fill-color);\n",
       "    border-color: transparent;\n",
       "    border-bottom-color: var(--fill-color);\n",
       "    animation:\n",
       "      spin 1s steps(1) infinite;\n",
       "  }\n",
       "\n",
       "  @keyframes spin {\n",
       "    0% {\n",
       "      border-color: transparent;\n",
       "      border-bottom-color: var(--fill-color);\n",
       "      border-left-color: var(--fill-color);\n",
       "    }\n",
       "    20% {\n",
       "      border-color: transparent;\n",
       "      border-left-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "    }\n",
       "    30% {\n",
       "      border-color: transparent;\n",
       "      border-left-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "      border-right-color: var(--fill-color);\n",
       "    }\n",
       "    40% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "    }\n",
       "    60% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "    }\n",
       "    80% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "      border-bottom-color: var(--fill-color);\n",
       "    }\n",
       "    90% {\n",
       "      border-color: transparent;\n",
       "      border-bottom-color: var(--fill-color);\n",
       "    }\n",
       "  }\n",
       "</style>\n",
       "\n",
       "  <script>\n",
       "    async function quickchart(key) {\n",
       "      const quickchartButtonEl =\n",
       "        document.querySelector('#' + key + ' button');\n",
       "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
       "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
       "      try {\n",
       "        const charts = await google.colab.kernel.invokeFunction(\n",
       "            'suggestCharts', [key], {});\n",
       "      } catch (error) {\n",
       "        console.error('Error during call to suggestCharts:', error);\n",
       "      }\n",
       "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
       "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
       "    }\n",
       "    (() => {\n",
       "      let quickchartButtonEl =\n",
       "        document.querySelector('#df-bcdcdbbe-db36-4ef8-af6d-40562fa763a1 button');\n",
       "      quickchartButtonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "    })();\n",
       "  </script>\n",
       "</div>\n",
       "\n",
       "  <div id=\"id_516ad04a-f7cf-4348-ac93-f5a5295596e2\">\n",
       "    <style>\n",
       "      .colab-df-generate {\n",
       "        background-color: #E8F0FE;\n",
       "        border: none;\n",
       "        border-radius: 50%;\n",
       "        cursor: pointer;\n",
       "        display: none;\n",
       "        fill: #1967D2;\n",
       "        height: 32px;\n",
       "        padding: 0 0 0 0;\n",
       "        width: 32px;\n",
       "      }\n",
       "\n",
       "      .colab-df-generate:hover {\n",
       "        background-color: #E2EBFA;\n",
       "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "        fill: #174EA6;\n",
       "      }\n",
       "\n",
       "      [theme=dark] .colab-df-generate {\n",
       "        background-color: #3B4455;\n",
       "        fill: #D2E3FC;\n",
       "      }\n",
       "\n",
       "      [theme=dark] .colab-df-generate:hover {\n",
       "        background-color: #434B5C;\n",
       "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "        fill: #FFFFFF;\n",
       "      }\n",
       "    </style>\n",
       "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('SARS1_df')\"\n",
       "            title=\"Generate code using this dataframe.\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "       width=\"24px\">\n",
       "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
       "  </svg>\n",
       "    </button>\n",
       "    <script>\n",
       "      (() => {\n",
       "      const buttonEl =\n",
       "        document.querySelector('#id_516ad04a-f7cf-4348-ac93-f5a5295596e2 button.colab-df-generate');\n",
       "      buttonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "      buttonEl.onclick = () => {\n",
       "        google.colab.notebook.generateWithVariable('SARS1_df');\n",
       "      }\n",
       "      })();\n",
       "    </script>\n",
       "  </div>\n",
       "\n",
       "    </div>\n",
       "  </div>\n"
      ],
      "text/plain": [
       "          CHROM    POS  ID REF ALT        QUAL  FILTER  GENOTYPE\n",
       "0   NC_045512.2     21 NaN   C   T   54.414600     NaN         3\n",
       "1   NC_045512.2    241 NaN   C   T  228.427994     NaN         3\n",
       "2   NC_045512.2   1387 NaN   C   T  225.421997     NaN         3\n",
       "3   NC_045512.2   2832 NaN   A   G  228.425995     NaN         3\n",
       "4   NC_045512.2   3037 NaN   C   T  225.417007     NaN         3\n",
       "5   NC_045512.2   5386 NaN   T   G  228.425995     NaN         3\n",
       "6   NC_045512.2   8393 NaN   G   A  225.417007     NaN         3\n",
       "7   NC_045512.2  10029 NaN   C   T  225.425995     NaN         3\n",
       "8   NC_045512.2  10449 NaN   C   A  201.427002     NaN         3\n",
       "9   NC_045512.2  11291 NaN   G   A   35.133900     NaN         3\n",
       "10  NC_045512.2  11537 NaN   A   G  225.425995     NaN         3\n",
       "11  NC_045512.2  13195 NaN   T   C  228.434998     NaN         3\n",
       "12  NC_045512.2  14408 NaN   C   T  225.417007     NaN         3\n",
       "13  NC_045512.2  15240 NaN   C   T  228.429001     NaN         3\n",
       "14  NC_045512.2  18163 NaN   A   G  225.417007     NaN         3\n",
       "15  NC_045512.2  21762 NaN   C   T   89.418999     NaN         3\n",
       "16  NC_045512.2  21846 NaN   C   T  225.425995     NaN         3\n",
       "17  NC_045512.2  22578 NaN   G   A  228.423996     NaN         3\n",
       "18  NC_045512.2  22673 NaN   T   C  228.434006     NaN         3\n",
       "19  NC_045512.2  22674 NaN   C   T  225.417007     NaN         3\n",
       "20  NC_045512.2  22679 NaN   T   C  225.421997     NaN         3\n",
       "21  NC_045512.2  22686 NaN   C   T  225.421997     NaN         3\n",
       "22  NC_045512.2  22813 NaN   G   T  228.408997     NaN         3\n",
       "23  NC_045512.2  22898 NaN   G   A  134.817993     NaN         3\n",
       "24  NC_045512.2  22992 NaN   G   A  225.421997     NaN         3\n",
       "25  NC_045512.2  22995 NaN   C   A  225.417007     NaN         3\n",
       "26  NC_045512.2  23013 NaN   A   C  225.425995     NaN         3\n",
       "27  NC_045512.2  23040 NaN   A   G  225.421997     NaN         3\n",
       "28  NC_045512.2  23048 NaN   G   A  225.425995     NaN         3\n",
       "29  NC_045512.2  23055 NaN   A   G  228.427002     NaN         3\n",
       "30  NC_045512.2  23063 NaN   A   T  225.421997     NaN         3\n",
       "31  NC_045512.2  23075 NaN   T   C  228.427002     NaN         3\n",
       "32  NC_045512.2  23202 NaN   C   A  228.419006     NaN         3\n",
       "33  NC_045512.2  23403 NaN   A   G  228.436005     NaN         3\n",
       "34  NC_045512.2  23525 NaN   C   T  225.417007     NaN         3\n",
       "35  NC_045512.2  23599 NaN   T   G  225.417007     NaN         3\n",
       "36  NC_045512.2  23604 NaN   C   A  225.421997     NaN         3\n",
       "37  NC_045512.2  23854 NaN   C   A  228.431000     NaN         3\n",
       "38  NC_045512.2  23948 NaN   G   T  225.421997     NaN         3\n",
       "39  NC_045512.2  24130 NaN   C   A  225.421997     NaN         3\n",
       "40  NC_045512.2  24424 NaN   A   T  228.425003     NaN         3\n",
       "41  NC_045512.2  24469 NaN   T   A  225.417007     NaN         3\n",
       "42  NC_045512.2  24503 NaN   C   T  228.427002     NaN         3\n",
       "43  NC_045512.2  25584 NaN   C   T  225.417007     NaN         3\n",
       "44  NC_045512.2  26270 NaN   C   T  225.425995     NaN         3\n",
       "45  NC_045512.2  26530 NaN   A   G  228.376007     NaN         3\n",
       "46  NC_045512.2  26709 NaN   G   A  225.425995     NaN         3\n",
       "47  NC_045512.2  27259 NaN   A   C  228.417007     NaN         3\n",
       "48  NC_045512.2  27807 NaN   C   T  228.416000     NaN         3\n",
       "49  NC_045512.2  28271 NaN   A   T  228.432999     NaN         3\n",
       "50  NC_045512.2  28311 NaN   C   T  228.429001     NaN         3\n",
       "51  NC_045512.2  28363 NaN   A   T   10.535200     NaN         3\n",
       "52  NC_045512.2  28370 NaN   A   G   76.404999     NaN         3\n",
       "53  NC_045512.2  28881 NaN   G   A  225.425995     NaN         3\n",
       "54  NC_045512.2  28882 NaN   G   A  228.436005     NaN         3\n",
       "55  NC_045512.2  28883 NaN   G   C  228.434006     NaN         3"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#use pandas to make dataframe for each genome\n",
    "import pandas\n",
    "columns = ['CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'GENOTYPE']\n",
    "SARS1_df = pandas.read_csv(\"output1.tsv\", delimiter = \"\\t\", header=None, names=columns)\n",
    "SARS1_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 394
    },
    "collapsed": true,
    "id": "WGdb1zHUBWu0",
    "outputId": "7660e365-2bf6-42b0-b85f-9ef733f74c33"
   },
   "outputs": [
    {
     "data": {
      "application/vnd.google.colaboratory.intrinsic+json": {
       "summary": "{\n  \"name\": \"SARS2_df\",\n  \"rows\": 11,\n  \"fields\": [\n    {\n      \"column\": \"CHROM\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"NC_045512.2\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"POS\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 9494,\n        \"min\": 241,\n        \"max\": 28854,\n        \"num_unique_values\": 11,\n        \"samples\": [\n          18292\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"ID\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": null,\n        \"min\": null,\n        \"max\": null,\n        \"num_unique_values\": 0,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"REF\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 3,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"ALT\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 3,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"QUAL\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 66.7021708759643,\n        \"min\": 5.756770133972168,\n        \"max\": 228.43600463867188,\n        \"num_unique_values\": 9,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"FILTER\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": null,\n        \"min\": null,\n        \"max\": null,\n        \"num_unique_values\": 0,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"GENOTYPE\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0,\n        \"min\": 3,\n        \"max\": 3,\n        \"num_unique_values\": 1,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}",
       "type": "dataframe",
       "variable_name": "SARS2_df"
      },
      "text/html": [
       "\n",
       "  <div id=\"df-7b57e1e9-a900-44bc-bb0b-382dd7e0ca4c\" class=\"colab-df-container\">\n",
       "    <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>GENOTYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>241</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>228.436005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>3037</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>225.417007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>11842</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>225.421997</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>12759</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>225.421997</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>14408</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>228.427002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>18292</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>225.425995</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>20268</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>225.425995</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>23403</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>228.432999</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>24978</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>5.756770</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>27813</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>228.425995</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>28854</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>228.431000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "    <div class=\"colab-df-buttons\">\n",
       "\n",
       "  <div class=\"colab-df-container\">\n",
       "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-7b57e1e9-a900-44bc-bb0b-382dd7e0ca4c')\"\n",
       "            title=\"Convert this dataframe to an interactive table.\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
       "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
       "  </svg>\n",
       "    </button>\n",
       "\n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    .colab-df-buttons div {\n",
       "      margin-bottom: 4px;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "    <script>\n",
       "      const buttonEl =\n",
       "        document.querySelector('#df-7b57e1e9-a900-44bc-bb0b-382dd7e0ca4c button.colab-df-convert');\n",
       "      buttonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "      async function convertToInteractive(key) {\n",
       "        const element = document.querySelector('#df-7b57e1e9-a900-44bc-bb0b-382dd7e0ca4c');\n",
       "        const dataTable =\n",
       "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                    [key], {});\n",
       "        if (!dataTable) return;\n",
       "\n",
       "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "          + ' to learn more about interactive tables.';\n",
       "        element.innerHTML = '';\n",
       "        dataTable['output_type'] = 'display_data';\n",
       "        await google.colab.output.renderOutput(dataTable, element);\n",
       "        const docLink = document.createElement('div');\n",
       "        docLink.innerHTML = docLinkHtml;\n",
       "        element.appendChild(docLink);\n",
       "      }\n",
       "    </script>\n",
       "  </div>\n",
       "\n",
       "\n",
       "<div id=\"df-0400520d-4f51-4078-85f6-95d71ba2f8e8\">\n",
       "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-0400520d-4f51-4078-85f6-95d71ba2f8e8')\"\n",
       "            title=\"Suggest charts\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "     width=\"24px\">\n",
       "    <g>\n",
       "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
       "    </g>\n",
       "</svg>\n",
       "  </button>\n",
       "\n",
       "<style>\n",
       "  .colab-df-quickchart {\n",
       "      --bg-color: #E8F0FE;\n",
       "      --fill-color: #1967D2;\n",
       "      --hover-bg-color: #E2EBFA;\n",
       "      --hover-fill-color: #174EA6;\n",
       "      --disabled-fill-color: #AAA;\n",
       "      --disabled-bg-color: #DDD;\n",
       "  }\n",
       "\n",
       "  [theme=dark] .colab-df-quickchart {\n",
       "      --bg-color: #3B4455;\n",
       "      --fill-color: #D2E3FC;\n",
       "      --hover-bg-color: #434B5C;\n",
       "      --hover-fill-color: #FFFFFF;\n",
       "      --disabled-bg-color: #3B4455;\n",
       "      --disabled-fill-color: #666;\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart {\n",
       "    background-color: var(--bg-color);\n",
       "    border: none;\n",
       "    border-radius: 50%;\n",
       "    cursor: pointer;\n",
       "    display: none;\n",
       "    fill: var(--fill-color);\n",
       "    height: 32px;\n",
       "    padding: 0;\n",
       "    width: 32px;\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart:hover {\n",
       "    background-color: var(--hover-bg-color);\n",
       "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "    fill: var(--button-hover-fill-color);\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart-complete:disabled,\n",
       "  .colab-df-quickchart-complete:disabled:hover {\n",
       "    background-color: var(--disabled-bg-color);\n",
       "    fill: var(--disabled-fill-color);\n",
       "    box-shadow: none;\n",
       "  }\n",
       "\n",
       "  .colab-df-spinner {\n",
       "    border: 2px solid var(--fill-color);\n",
       "    border-color: transparent;\n",
       "    border-bottom-color: var(--fill-color);\n",
       "    animation:\n",
       "      spin 1s steps(1) infinite;\n",
       "  }\n",
       "\n",
       "  @keyframes spin {\n",
       "    0% {\n",
       "      border-color: transparent;\n",
       "      border-bottom-color: var(--fill-color);\n",
       "      border-left-color: var(--fill-color);\n",
       "    }\n",
       "    20% {\n",
       "      border-color: transparent;\n",
       "      border-left-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "    }\n",
       "    30% {\n",
       "      border-color: transparent;\n",
       "      border-left-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "      border-right-color: var(--fill-color);\n",
       "    }\n",
       "    40% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "    }\n",
       "    60% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "    }\n",
       "    80% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "      border-bottom-color: var(--fill-color);\n",
       "    }\n",
       "    90% {\n",
       "      border-color: transparent;\n",
       "      border-bottom-color: var(--fill-color);\n",
       "    }\n",
       "  }\n",
       "</style>\n",
       "\n",
       "  <script>\n",
       "    async function quickchart(key) {\n",
       "      const quickchartButtonEl =\n",
       "        document.querySelector('#' + key + ' button');\n",
       "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
       "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
       "      try {\n",
       "        const charts = await google.colab.kernel.invokeFunction(\n",
       "            'suggestCharts', [key], {});\n",
       "      } catch (error) {\n",
       "        console.error('Error during call to suggestCharts:', error);\n",
       "      }\n",
       "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
       "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
       "    }\n",
       "    (() => {\n",
       "      let quickchartButtonEl =\n",
       "        document.querySelector('#df-0400520d-4f51-4078-85f6-95d71ba2f8e8 button');\n",
       "      quickchartButtonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "    })();\n",
       "  </script>\n",
       "</div>\n",
       "\n",
       "  <div id=\"id_1b91b339-fe14-4b35-a501-4c5468e1c280\">\n",
       "    <style>\n",
       "      .colab-df-generate {\n",
       "        background-color: #E8F0FE;\n",
       "        border: none;\n",
       "        border-radius: 50%;\n",
       "        cursor: pointer;\n",
       "        display: none;\n",
       "        fill: #1967D2;\n",
       "        height: 32px;\n",
       "        padding: 0 0 0 0;\n",
       "        width: 32px;\n",
       "      }\n",
       "\n",
       "      .colab-df-generate:hover {\n",
       "        background-color: #E2EBFA;\n",
       "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "        fill: #174EA6;\n",
       "      }\n",
       "\n",
       "      [theme=dark] .colab-df-generate {\n",
       "        background-color: #3B4455;\n",
       "        fill: #D2E3FC;\n",
       "      }\n",
       "\n",
       "      [theme=dark] .colab-df-generate:hover {\n",
       "        background-color: #434B5C;\n",
       "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "        fill: #FFFFFF;\n",
       "      }\n",
       "    </style>\n",
       "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('SARS2_df')\"\n",
       "            title=\"Generate code using this dataframe.\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "       width=\"24px\">\n",
       "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
       "  </svg>\n",
       "    </button>\n",
       "    <script>\n",
       "      (() => {\n",
       "      const buttonEl =\n",
       "        document.querySelector('#id_1b91b339-fe14-4b35-a501-4c5468e1c280 button.colab-df-generate');\n",
       "      buttonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "      buttonEl.onclick = () => {\n",
       "        google.colab.notebook.generateWithVariable('SARS2_df');\n",
       "      }\n",
       "      })();\n",
       "    </script>\n",
       "  </div>\n",
       "\n",
       "    </div>\n",
       "  </div>\n"
      ],
      "text/plain": [
       "          CHROM    POS  ID REF ALT        QUAL  FILTER  GENOTYPE\n",
       "0   NC_045512.2    241 NaN   C   T  228.436005     NaN         3\n",
       "1   NC_045512.2   3037 NaN   C   T  225.417007     NaN         3\n",
       "2   NC_045512.2  11842 NaN   G   A  225.421997     NaN         3\n",
       "3   NC_045512.2  12759 NaN   A   G  225.421997     NaN         3\n",
       "4   NC_045512.2  14408 NaN   C   T  228.427002     NaN         3\n",
       "5   NC_045512.2  18292 NaN   G   T  225.425995     NaN         3\n",
       "6   NC_045512.2  20268 NaN   A   G  225.425995     NaN         3\n",
       "7   NC_045512.2  23403 NaN   A   G  228.432999     NaN         3\n",
       "8   NC_045512.2  24978 NaN   A   T    5.756770     NaN         3\n",
       "9   NC_045512.2  27813 NaN   C   T  228.425995     NaN         3\n",
       "10  NC_045512.2  28854 NaN   C   T  228.431000     NaN         3"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns = ['CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'GENOTYPE']\n",
    "SARS2_df = pandas.read_csv(\"output2.tsv\", delimiter = \"\\t\", header=None, names=columns)\n",
    "SARS2_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "3qPsFP4pBkjb",
    "outputId": "71e0db31-ab95-493d-963a-cea531559a64"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Same Mutations:\n",
      "     POS REF_1 ALT_1 REF_2 ALT_2\n",
      "0    241     C     T     C     T\n",
      "1   3037     C     T     C     T\n",
      "2  14408     C     T     C     T\n",
      "3  23403     A     G     A     G\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# combined the two dataframes on the 'Position' column to compare mutations at the same positions\n",
    "combined_df = pd.merge(SARS1_df, SARS2_df, on= 'POS', suffixes=('_1', '_2'))\n",
    "\n",
    "# Find positions where mutations are the same in both datasets\n",
    "same_mutations = combined_df[combined_df['REF_1'] == combined_df['REF_2']]\n",
    "same_mutations = same_mutations[same_mutations['ALT_1'] == same_mutations['ALT_2']]\n",
    "\n",
    "# Output results\n",
    "print(\"Same Mutations:\")\n",
    "print(same_mutations[['POS', 'REF_1', 'ALT_1', 'REF_2', 'ALT_2']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 426
    },
    "id": "ZuvcG4H8CTan",
    "outputId": "770152e3-bd45-4d85-8adb-e0546709b7c7"
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0kAAAIjCAYAAADWYVDIAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAABKu0lEQVR4nO3dd3RU1f7+8WcSSAgkhJJACIYk9F4FRES69KZcQblShasEkaIg10tXURSkg9evEkRFRKRYADG0Sy+ChQ6GotTQQgIkkOzfH67M74wpZHCSCfH9WmvWYvbZZ5/PTE7Kwz5nj80YYwQAAAAAkCR5uLsAAAAAAMhJCEkAAAAAYEFIAgAAAAALQhIAAAAAWBCSAAAAAMCCkAQAAAAAFoQkAAAAALAgJAEAAACABSEJAAAAACwISQDuG+PGjZPNZsuWYzVp0kRNmjSxP9+wYYNsNpu++OKLbDl+7969FRYWli3HuldxcXF69tlnFRQUJJvNpiFDhri7pAylfA03bNhgb7sf3uf7WVxcnIoVK6ZPPvnE3aVki+7du+vJJ590dxkAXICQBMAtIiMjZbPZ7I98+fIpODhYrVq10owZM3T9+nWXHOfMmTMaN26c9u3b55LxXCkn15YZb7zxhiIjI/X8889r4cKFeuaZZzLsf/v2bc2YMUN169aVn5+ffH19VbduXc2cOVN37tzJpqozduPGDY0bN84hSLnKzz//rK5duyo0NFT58uVTyZIl1bJlS82cOdPlx8oppk+fLj8/P3Xv3t2hffPmzWrTpo1KliypfPnyqVSpUurQoYM+/fTTdMeqV6+ebDab5s6dm+b2P/9MyZMnj0qWLKnevXvr999/T9U/OTlZH330kerXr68iRYrIz89P5cuXV8+ePbV9+3Z7v0OHDmnEiBGqWbOm/Pz8VKJECbVr1067d+9ONebIkSO1dOlS/fjjj5l9iwDkUDZjjHF3EQD+fiIjI9WnTx9NmDBB4eHhun37ts6dO6cNGzZo7dq1KlWqlFauXKnq1avb97lz547u3LmjfPnyZfo4u3fvVt26dTV//nz17t070/slJiZKkry8vCT9MQvRtGlTLVmyRF27ds30OPda2+3bt5WcnCxvb2+XHCsrPPTQQ8qTJ482b958177x8fFq166dNm7cqPbt26t169by8PDQ6tWrtXLlSjVr1kxfffWV8ufPn2X1pnwN169fb58l/PP7HBMTo8DAQI0dO1bjxo1z2bG3bt2qpk2bqlSpUurVq5eCgoJ0+vRpbd++XcePH9exY8dcdqyc4vbt2ypZsqSGDh2qUaNG2duXLFmibt26qWbNmurevbsKFy6s6Ohobdq0SXnz5tX69etTjXX06FGVL19eYWFhKlmyZJrn3J9/pty6dUvbt29XZGSkwsLC9Msvvzj87Bg0aJBmz56tTp06qVmzZsqTJ48OHz6sVatW6emnn7Z//V966SV98MEHeuKJJ1SvXj1du3ZN7733nk6cOKHVq1erRYsWDnXUr19fFSpU0EcffeSidxKAWxgAcIP58+cbSWbXrl2ptkVFRRkfHx8TGhpqbty48ZeOs2vXLiPJzJ8/P1P94+Pj02xfv369kWSWLFnyl+r5K7XlNOHh4aZdu3aZ6jtgwAAjycycOTPVtlmzZhlJZuDAga4u0UHK13D9+vXp9rl48aKRZMaOHevSY7dt29YEBgaaK1eupNp2/vx5lx7rbtI7x13tyy+/NJLMsWPHHNorV65sqlSpYhISElLtk957MWbMGFOsWDGzdOlSY7PZTHR0dKo+6f1MGTlypJFkFi9ebG87d+6csdlspn///qnGSU5Odqhj9+7d5vr16w59YmJiTGBgoGnYsGGq/d955x1ToECBVPsAuL9wuR2AHKdZs2YaPXq0Tp48qY8//tjentY9SWvXrtUjjzyiQoUKydfXVxUqVNC///1vSX/MHNStW1eS1KdPH/tlOJGRkZL+uO+oatWq2rNnjx599FHlz5/fvu+f70lKkZSUpH//+98KCgpSgQIF1LFjR50+fdqhT1hYWJqzVtYx71ZbWvfKxMfHa/jw4QoJCZG3t7cqVKigd955R+ZPFwTYbDYNGjRIy5cvV9WqVeXt7a0qVapo9erVab/hf3LhwgX169dPxYsXV758+VSjRg0tWLDAvj3l3p7o6Gh988039tpPnDiR5ni//fabPvjgAzVr1kyDBg1KtT0iIkJNmzbVf//7X/tlUSdOnHB4P/78+qyzPCdPntTAgQNVoUIF+fj4qGjRovrHP/6Rbj1W1vf5xIkTCgwMlCSNHz/e/rrGjRun+fPny2azae/evanGeOONN+Tp6ZnmJV0pjh8/ripVqqhQoUKpthUrVixV28cff6x69eopf/78Kly4sB599FF99913Dn3mzJmjKlWqyNvbW8HBwYqIiNDVq1cd+mR0jickJGjs2LEqW7asvL29FRISohEjRighIcFhjIy+xzKyfPlyhYWFqUyZMqnei7p169pnae/2XkjSp59+qq5du6p9+/by9/fP8LK8P2vUqJH9uCmio6NljFHDhg1T9bfZbA511KlTR76+vg59ihYtqkaNGungwYOp9m/ZsqXi4+O1du3aTNcIIOchJAHIkVLub/nzH4ZW+/fvV/v27ZWQkKAJEyZoypQp6tixo7Zs2SJJqlSpkiZMmCBJGjBggBYuXKiFCxfq0UcftY9x6dIltWnTRjVr1tS0adPUtGnTDOt6/fXX9c0332jkyJEaPHiw1q5dqxYtWujmzZtOvb7M1GZljFHHjh317rvvqnXr1po6daoqVKigl19+WcOGDUvVf/PmzRo4cKC6d++uyZMn69atW3riiSd06dKlDOu6efOmmjRpooULF6pHjx56++235e/vr969e2v69On22hcuXKiAgADVrFnTXntKwPizVatWKSkpST179kz3uD179tSdO3cyHeSsdu3apa1bt6p79+6aMWOGnnvuOUVFRalJkya6ceNGpscJDAy03+/SpUsX++t6/PHH1bVrV/n4+KS5AMEnn3yiJk2aqGTJkumOHRoaqj179uiXX365ax3jx4/XM888o7x582rChAkaP368QkJCtG7dOnufcePGKSIiQsHBwZoyZYqeeOIJvffee3rsscd0+/Zth/HSOseTk5PVsWNHvfPOO+rQoYNmzpypzp07691331W3bt3s+97teywjW7duVe3atdN8L6KiovTbb7/ddQxJ2rFjh44dO6annnpKXl5eevzxx51aCCIlLBcuXNihBumPS/+cOUeszp07p4CAgFTtlStXlo+PT6beIwA5mJtnsgD8TWV0uV0Kf39/U6tWLfvzsWPHGuuPrXfffddIMhcvXkx3jIwuaWvcuLGRZObNm5fmtsaNG9ufp1yqVbJkSRMbG2tv//zzz40kM336dHtbaGio6dWr113HzKi2Xr16mdDQUPvz5cuXG0nmtddec+jXtWtXY7PZHC5pkmS8vLwc2n788cd0L3ezmjZtmpFkPv74Y3tbYmKiadCggfH19XV47aGhoZm63G7IkCFGktm7d2+6fX744QcjyQwbNswYY0x0dHS6743+dDlcWpdkbtu2zUgyH330kb0trcvt/vw+Z3S53VNPPWWCg4NNUlJSqrrvdsnkd999Zzw9PY2np6dp0KCBGTFihFmzZo1JTEx06Hf06FHj4eFhunTp4nAcY/64DMwYYy5cuGC8vLzMY4895tAn5bLFDz/80N6W3jm+cOFC4+HhYf73v/85tM+bN89IMlu2bDHGZO57LC23b982NpvNDB8+PNW2Dz74wH6ONm3a1IwePdr873//S/V6UwwaNMiEhITYX/93332X5vmU8jPl+++/NxcvXjSnT582X3zxhQkMDDTe3t7m9OnTDv179uxpJJnChQubLl26mHfeecccPHgwU69v06ZNxmazmdGjR6e5vXz58qZNmzaZGgtAzsRMEoAcy9fXN8NV7lIuXVqxYoWSk5Pv6Rje3t7q06dPpvv37NlTfn5+9uddu3ZViRIl9O23397T8TPr22+/laenpwYPHuzQPnz4cBljtGrVKof2Fi1aOFzmVL16dRUsWFC//vrrXY8TFBSkp556yt6WN29eDR48WHFxcdq4caPTtad8Da3v25+lbLuXVQ19fHzs/759+7YuXbqksmXLqlChQvrhhx+cHi89PXv21JkzZxwWFvjkk0/k4+OjJ554IsN9W7ZsqW3btqljx4768ccfNXnyZLVq1UolS5bUypUr7f2WL1+u5ORkjRkzRh4ejr+iUy41/f7775WYmKghQ4Y49Onfv78KFiyob775xmG/tM7xJUuWqFKlSqpYsaJiYmLsj2bNmkmS/TXe6/fY5cuXZYxxmL1J0bdvX61evVpNmjTR5s2bNXHiRDVq1EjlypXT1q1bHfreuXNHixcvVrdu3eyvv1mzZhkuK96iRQsFBgYqJCREXbt2VYECBbRy5Uo98MADDv3mz5+vWbNmKTw8XMuWLdNLL72kSpUqqXnz5hleOnnhwgU9/fTTCg8P14gRI9LsU7hwYcXExGT4HgHI2QhJAHKsuLi4DP+w7tatmxo2bKhnn31WxYsXV/fu3fX555879cdcyZIl07w3Ij3lypVzeG6z2VS2bNlM3f/yV5w8eVLBwcGp3o9KlSrZt1uVKlUq1RiFCxfWlStX7nqccuXKpfoDPb3jZEZmAlDKtvTuScnIzZs3NWbMGPu9WgEBAQoMDNTVq1d17do1p8dLT8uWLVWiRAn7H+fJyclatGiROnXqlOF5mqJu3br68ssvdeXKFe3cuVOjRo3S9evX1bVrVx04cEDSH/fNeHh4qHLlyumOk/I1qFChgkO7l5eXSpcuneprlNY5fvToUe3fv1+BgYEOj/Lly0v6IwhIf/17zKSzgG6rVq20Zs0aXb16VZs2bVJERIROnjyp9u3b248t/XG57cWLF1WvXj0dO3ZMx44dU3R0tJo2bapFixalWcfs2bO1du1affHFF2rbtq1iYmLSXCXSw8NDERER2rNnj2JiYrRixQq1adNG69atS7VkeYr4+Hi1b99e169f14oVK1Ldq2R93dn1mW4AskYedxcAAGn57bffdO3aNZUtWzbdPj4+Ptq0aZPWr1+vb775RqtXr9bixYvVrFkzfffdd/L09LzrcayzEK6S3h9HSUlJmarJFdI7Tnp/tGallD/4f/rpJ9WsWTPNPj/99JMkqXTp0pIyfg//7IUXXtD8+fM1ZMgQNWjQQP7+/rLZbOrevfs9zzCmxdPTU08//bTef/99zZkzR1u2bNGZM2f0z3/+06lxvLy8VLduXdWtW1fly5dXnz59tGTJEo0dO9ZltVqldY4nJyerWrVqmjp1apr7hISE2Pe9l++xIkWKyGaz3TWU58+fX40aNVKjRo0UEBCg8ePHa9WqVerVq5ck2QNpeh/QunHjxlT3EdarV08PPvigJKlz58565JFH9PTTT+vw4cPphpqiRYuqY8eO6tixo5o0aaKNGzfq5MmT9nuXpD8+FuDxxx/XTz/9pDVr1qhq1arpvq4rV66k+g8VAPcXZpIA5EgLFy6U9Mf/OGfEw8NDzZs319SpU3XgwAG9/vrrWrdunf1yIVf/b+7Ro0cdnhtjdOzYMYeV6AoXLpxqlTEp9SyMM7WFhobqzJkzqWZjDh06ZN/uCqGhoTp69GiqcPFXjtOmTRt5enrav6Zp+eijj+Tl5aVOnTpJ+v832f/5fUxrJuuLL75Qr169NGXKFHXt2lUtW7bUI488kubX4G7u9jXp2bOnYmNj9dVXX+mTTz5RYGDgXc/RjKT8MX/27FlJUpkyZZScnGyfWUpLytfg8OHDDu2JiYmKjo7O1NeoTJkyunz5spo3b64WLVqkelhnqe72PZaWPHnyqEyZMoqOjr5rLSn+/F7Ex8drxYoV6tatm5YsWZLqYZ3VS4+np6cmTZqkM2fOaNasWfdUh/RHqOzZs6eioqL06aefqnHjxunuf+fOHZ0+fdo++wrg/kRIApDjrFu3ThMnTlR4eLh69OiRbr/Lly+nakuZqUhZxrhAgQKSUv+xfa8++ugjh6DyxRdf6OzZs2rTpo29rUyZMtq+fbv9A2kl6euvv061VLgztbVt21ZJSUmp/tB79913ZbPZHI7/V7Rt21bnzp3T4sWL7W137tzRzJkz5evrm+Efh+l54IEH1K9fP33//ff21eOs5s2bp3Xr1ulf//qXihYtKkkqWLCgAgICtGnTJoe+c+bMSbW/p6dnqhmymTNnpjnrdDcpH2ab3tekevXqql69uv7v//5PS5cuVffu3ZUnz90vyli/fn2as3gp97KlhJLOnTvLw8NDEyZMSBVUU/Zv0aKFvLy8NGPGDIcxP/jgA127dk3t2rW7az1PPvmkfv/9d73//vuptt28eVPx8fGSMvc9lp4GDRpo9+7dqdqjoqLS7P/n92LZsmWKj49XRESEunbtmurRvn17LV269K51NGnSRPXq1dO0adN069YtSX+sTJdWEE1MTFRUVJQ8PDwcZrFfeOEFLV68WHPmzNHjjz+e4fEOHDigW7du6eGHH86wH4CcjcvtALjVqlWrdOjQId25c0fnz5/XunXrtHbtWoWGhmrlypXKly9fuvtOmDBBmzZtUrt27RQaGqoLFy5ozpw5euCBB/TII49I+iOwFCpUSPPmzZOfn58KFCig+vXrKzw8/J7qLVKkiB555BH16dNH58+f17Rp01S2bFn179/f3ufZZ5/VF198odatW+vJJ5/U8ePH9fHHH6f6vBhnauvQoYOaNm2qV199VSdOnFCNGjX03XffacWKFRoyZEiqse/VgAED9N5776l3797as2ePwsLC9MUXX2jLli2aNm1apu69ScvUqVN16NAhDRw4UKtXr1br1q0lSWvWrNGKFSvUrFkzvf322w77PPvss3rzzTf17LPP6sEHH9SmTZt05MiRVGO3b99eCxculL+/vypXrqxt27bp+++/twcuZ/j4+Khy5cpavHixypcvryJFiqhq1aoOl1b17NlTL730kiRl+lK7F154QTdu3FCXLl1UsWJFJSYmauvWrVq8eLHCwsLsCyuULVtWr776qn0xg8cff1ze3t7atWuXgoODNWnSJAUGBmrUqFEaP368WrdurY4dO+rw4cOaM2eO6tatm6mannnmGX3++ed67rnntH79ejVs2FBJSUk6dOiQPv/8c61Zs0YPPvhgpr7H0tOpUyctXLhQR44csd/rlNIeHh6uDh06qEyZMoqPj9f333+vr776SnXr1lWHDh0k/XGpXdGiRdMNGx07dtT777+vb7755q7B5eWXX9Y//vEPRUZG6rnnntNvv/2mevXqqVmzZmrevLmCgoJ04cIFLVq0SD/++KOGDBliX9572rRpmjNnjho0aKD8+fM7fHab9Mdy8Sn/4SH98blS+fPnV8uWLTOsCUAO565l9QD8vaUs15vy8PLyMkFBQaZly5Zm+vTpDktNp/jzEuBRUVGmU6dOJjg42Hh5eZng4GDz1FNPmSNHjjjst2LFClO5cmWTJ08eh+WaGzdubKpUqZJmfektAb5o0SIzatQoU6xYMePj42PatWtnTp48mWr/KVOmmJIlSxpvb2/TsGFDs3v37lRjZlTbn5emNsaY69evm6FDh5rg4GCTN29eU65cOfP222/bl0ZOIclERESkqim9pcn/7Pz586ZPnz4mICDAeHl5mWrVqqW5xHVmlwBPkZiYaKZNm2bq1Klj8ufPb//a9+rVK83ln2/cuGH69etn/P39jZ+fn3nyySfNhQsXUi3RfeXKFXu9vr6+plWrVubQoUOpXm9mlgA3xpitW7eaOnXqGC8vrzSXAz979qzx9PQ05cuXz/RrX7Vqlenbt6+pWLGi8fX1NV5eXqZs2bLmhRdeMOfPn0/V/8MPPzS1atUy3t7epnDhwqZx48Zm7dq1Dn1mzZplKlasaPLmzWuKFy9unn/+eXPlyhWHPhmd44mJieatt94yVapUsR+nTp06Zvz48ebatWvGmMx/j6UlISHBBAQEmIkTJzq0L1q0yHTv3t2UKVPG+Pj4mHz58pnKlSubV1991f59f/78eZMnTx7zzDPPpDv+jRs3TP78+U2XLl2MMRl/rEBSUpIpU6aMKVOmjLlz546JjY0106dPN61atTIPPPCAyZs3r/Hz8zMNGjQw77//vsP3VK9evRx+Vv35ER0d7XCs+vXrm3/+8593fX8A5Gw2Y9xwFy8A4G8vNjZWjRs31vHjx7Vp06Z0F3XIaWJiYlSiRAmNGTNGo0ePdnc5OdrEiRM1f/58HT16NNsWLXGnffv2qXbt2vrhhx/um/MZQNq4JwkA4BYFCxbUqlWrFBAQoLZt297T8uLuEBkZqaSkJD3zzDPuLiXHGzp0qOLi4vTZZ5+5u5Rs8eabb6pr164EJCAXYCYJAIBMWLdunQ4cOKDRo0eradOm+vLLL91dEgAgixCSAADIhCZNmmjr1q1q2LChPv74Y5UsWdLdJQEAsgghCQAAAAAsuCcJAAAAACwISQAAAABgkes/TDY5OVlnzpyRn5+fbDabu8sBAAAA4CbGGF2/fl3BwcHy8Eh/vijXh6QzZ84oJCTE3WUAAAAAyCFOnz6tBx54IN3tuT4k+fn5SfrjjShYsKCbqwEAAADgLrGxsQoJCbFnhPTk+pCUcoldwYIFCUkAAAAA7nobDgs3AAAAAIAFIQkAAAAALAhJAAAAAGBBSAIAAAAAC0ISAAAAAFgQkgAAAADAgpAEAAAAABaEJAAAAACwICQBAAAAgAUhCQAAAAAsCEkAAAAAYEFIAgAAAAALQhIAAAAAWBCSAAAAAMCCkAQAAAAAFm4NSZs2bVKHDh0UHBwsm82m5cuXO2w3xmjMmDEqUaKEfHx81KJFCx09etQ9xQIAAAD4W3BrSIqPj1eNGjU0e/bsNLdPnjxZM2bM0Lx587Rjxw4VKFBArVq10q1bt7K5UgAAAAB/F3ncefA2bdqoTZs2aW4zxmjatGn6z3/+o06dOkmSPvroIxUvXlzLly9X9+7ds7NUAAAAAH8Tbg1JGYmOjta5c+fUokULe5u/v7/q16+vbdu2pRuSEhISlJCQYH8eGxub5bUCAAAA97NTp04pJiYmy8YPCAhQqVKlsmx8V8uxIencuXOSpOLFizu0Fy9e3L4tLZMmTdL48eOztDYAAAAgtzh16pQqVKykWzdvZNkx8vnk1+FDB++boJRjQ9K9GjVqlIYNG2Z/Hhsbq5CQEDdWBAAAAORcMTExunXzhoq2H668RV3/d/PtS6d16espiomJIST9VUFBQZKk8+fPq0SJEvb28+fPq2bNmunu5+3tLW9v76wuDwAAAMhV8hYNkXdQWXeXkSPk2M9JCg8PV1BQkKKiouxtsbGx2rFjhxo0aODGygAAAADkZm6dSYqLi9OxY8fsz6Ojo7Vv3z4VKVJEpUqV0pAhQ/Taa6+pXLlyCg8P1+jRoxUcHKzOnTu7r2gAAAAAuZpbQ9Lu3bvVtGlT+/OUe4l69eqlyMhIjRgxQvHx8RowYICuXr2qRx55RKtXr1a+fPncVTIAAACAXM6tIalJkyYyxqS73WazacKECZowYUI2VgUAAADg7yzH3pMEAAAAAO5ASAIAAAAAC0ISAAAAAFgQkgAAAADAgpAEAAAAABaEJAAAAACwICQBAAAAgAUhCQAAAAAsCEkAAAAAYEFIAgAAAAALQhIAAAAAWBCSAAAAAMCCkAQAAAAAFoQkAAAAALAgJAEAAACABSEJAAAAACwISQAAAABgQUgCAAAAAAtCEgAAAABYEJIAAAAAwIKQBAAAAAAWhCQAAAAAsCAkAQAAAIAFIQkAAAAALAhJAAAAAGBBSAIAAAAAC0ISAAAAAFgQkgAAAADAgpAEAAAAABaEJAAAAACwICQBAAAAgAUhCQAAAAAsCEkAAAAAYEFIAgAAAAALQhIAAAAAWBCSAAAAAMCCkAQAAAAAFoQkAAAAALAgJAEAAACABSEJAAAAACwISQAAAABgQUgCAAAAAAtCEgAAAABYEJIAAAAAwIKQBAAAAAAWhCQAAAAAsCAkAQAAAIAFIQkAAAAALAhJAAAAAGBBSAIAAAAAC0ISAAAAAFgQkgAAAADAgpAEAAAAABaEJAAAAACwICQBAAAAgAUhCQAAAAAsCEkAAAAAYEFIAgAAAAALQhIAAAAAWBCSAAAAAMCCkAQAAAAAFoQkAAAAALAgJAEAAACABSEJAAAAACwISQAAAABgQUgCAAAAAAtCEgAAAABYEJIAAAAAwIKQBAAAAAAWhCQAAAAAsCAkAQAAAIAFIQkAAAAALAhJAAAAAGBBSAIAAAAAC0ISAAAAAFgQkgAAAADAgpAEAAAAABY5OiQlJSVp9OjRCg8Pl4+Pj8qUKaOJEyfKGOPu0gAAAADkUnncXUBG3nrrLc2dO1cLFixQlSpVtHv3bvXp00f+/v4aPHiwu8sDAAAAkAvl6JC0detWderUSe3atZMkhYWFadGiRdq5c6ebKwMAAACQW+Xoy+0efvhhRUVF6ciRI5KkH3/8UZs3b1abNm3S3SchIUGxsbEODwAAAADIrBw9k/TKK68oNjZWFStWlKenp5KSkvT666+rR48e6e4zadIkjR8/PhurBAAAAJCb5OiZpM8//1yffPKJPv30U/3www9asGCB3nnnHS1YsCDdfUaNGqVr167ZH6dPn87GigEAAADc73L0TNLLL7+sV155Rd27d5ckVatWTSdPntSkSZPUq1evNPfx9vaWt7d3dpYJAAAAIBfJ0TNJN27ckIeHY4menp5KTk52U0UAAAAAcrscPZPUoUMHvf766ypVqpSqVKmivXv3aurUqerbt6+7SwMAAACQS+XokDRz5kyNHj1aAwcO1IULFxQcHKx//etfGjNmjLtLAwAAAJBL5eiQ5Ofnp2nTpmnatGnuLgUAAADA30SOvicJAAAAALIbIQkAAAAALAhJAAAAAGBBSAIAAAAAC0ISAAAAAFgQkgAAAADAgpAEAAAAABaEJAAAAACwICQBAAAAgAUhCQAAAAAsCEkAAAAAYEFIAgAAAAALQhIAAAAAWBCSAAAAAMCCkAQAAAAAFoQkAAAAALAgJAEAAACABSEJAAAAACwISQAAAABgQUgCAAAAAAtCEgAAAABYEJIAAAAAwIKQBAAAAAAWhCQAAAAAsCAkAQAAAIAFIQkAAAAALAhJAAAAAGBBSAIAAAAAC0ISAAAAAFgQkgAAAADAgpAEAAAAABaEJAAAAACwICQBAAAAgAUhCQAAAAAsCEkAAAAAYEFIAgAAAAALQhIAAAAAWBCSAAAAAMCCkAQAAAAAFoQkAAAAALAgJAEAAACABSEJAAAAACwISQAAAABg4XRIWrBggb755hv78xEjRqhQoUJ6+OGHdfLkSZcWBwAAAADZzemQ9MYbb8jHx0eStG3bNs2ePVuTJ09WQECAhg4d6vICAQAAACA75XF2h9OnT6ts2bKSpOXLl+uJJ57QgAED1LBhQzVp0sTV9QEAAABAtnJ6JsnX11eXLl2SJH333Xdq2bKlJClfvny6efOma6sDAAAAgGzm9ExSy5Yt9eyzz6pWrVo6cuSI2rZtK0nav3+/wsLCXF0fAAAAAGQrp2eSZs+erQYNGujixYtaunSpihYtKknas2ePnnrqKZcXCAAAAADZyemZpEKFCmnWrFmp2sePH++SggAAAADAnZwOSZJ09epV7dy5UxcuXFBycrK93Waz6ZlnnnFZcQAAAACQ3ZwOSV999ZV69OihuLg4FSxYUDabzb6NkAQAAADgfuf0PUnDhw9X3759FRcXp6tXr+rKlSv2x+XLl7OiRgAAAADINk6HpN9//12DBw9W/vz5s6IeAAAAAHArp0NSq1attHv37qyoBQAAAADczul7ktq1a6eXX35ZBw4cULVq1ZQ3b16H7R07dnRZcQAAAACQ3ZwOSf3795ckTZgwIdU2m82mpKSkv14VAAAAALiJ0yHJuuQ3AAAAAOQ2Tt+TBAAAAAC52T2FpI0bN6pDhw4qW7asypYtq44dO+p///ufq2sDAAAAgGzndEj6+OOP1aJFC+XPn1+DBw/W4MGD5ePjo+bNm+vTTz/NihoBAAAAINs4fU/S66+/rsmTJ2vo0KH2tsGDB2vq1KmaOHGinn76aZcWCAAAAADZyemZpF9//VUdOnRI1d6xY0dFR0e7pCgAAAAAcBenQ1JISIiioqJStX///fcKCQlxSVEAAAAA4C5OX243fPhwDR48WPv27dPDDz8sSdqyZYsiIyM1ffp0lxcIAAAAANnJ6ZD0/PPPKygoSFOmTNHnn38uSapUqZIWL16sTp06ubxAAAAAAMhOTockSerSpYu6dOni6loAAAAAwO34MFkAAAAAsMjUTFKRIkV05MgRBQQEqHDhwrLZbOn2vXz5ssuKAwAAAIDslqmQ9O6778rPz8/+74xCEgAAAADczzIVknr16mX/d+/evbOqFgAAAABwO6fvSfL09NSFCxdStV+6dEmenp4uKQoAAAAA3MXpkGSMSbM9ISFBXl5ef7kgAAAAAHCnTC8BPmPGDEmSzWbT//3f/8nX19e+LSkpSZs2bVLFihVdXyEAAAAAZKNMh6R3331X0h8zSfPmzXO4tM7Ly0thYWGaN2+e6ysEAAAAgGyU6ZAUHR0tSWratKm+/PJLFS5cOMuKAgAAAAB3yXRISrF+/fqsqAMAAAAAcgSnQ5Ik/fbbb1q5cqVOnTqlxMREh21Tp051SWEpfv/9d40cOVKrVq3SjRs3VLZsWc2fP18PPvigS48DAAAAANI9hKSoqCh17NhRpUuX1qFDh1S1alWdOHFCxhjVrl3bpcVduXJFDRs2VNOmTbVq1SoFBgbq6NGjXOoHAAAAIMs4HZJGjRqll156SePHj5efn5+WLl2qYsWKqUePHmrdurVLi3vrrbcUEhKi+fPn29vCw8NdegwAAAAAsHL6c5IOHjyonj17SpLy5MmjmzdvytfXVxMmTNBbb73l0uJWrlypBx98UP/4xz9UrFgx1apVS++//36G+yQkJCg2NtbhAQAAAACZ5XRIKlCggP0+pBIlSuj48eP2bTExMa6rTNKvv/6quXPnqly5clqzZo2ef/55DR48WAsWLEh3n0mTJsnf39/+CAkJcWlNAAAAAHI3py+3e+ihh7R582ZVqlRJbdu21fDhw/Xzzz/ryy+/1EMPPeTS4pKTk/Xggw/qjTfekCTVqlVLv/zyi+bNm6devXqluc+oUaM0bNgw+/PY2FiCEgAAAIBMczokTZ06VXFxcZKk8ePHKy4uTosXL1a5cuVcvrJdiRIlVLlyZYe2SpUqaenSpenu4+3tLW9vb5fWAQAAAODvw+mQVLp0afu/CxQooHnz5rm0IKuGDRvq8OHDDm1HjhxRaGholh0TAAAAwN+b0/cklS5dWpcuXUrVfvXqVYcA5QpDhw7V9u3b9cYbb+jYsWP69NNP9d///lcREREuPQ4AAAAApHA6JJ04cUJJSUmp2hMSEvT777+7pKgUdevW1bJly7Ro0SJVrVpVEydO1LRp09SjRw+XHgcAAAAAUmT6cruVK1fa/71mzRr5+/vbnyclJSkqKkphYWEuLU6S2rdvr/bt27t8XAAAAABIS6ZDUufOnSVJNpst1cpyefPmVVhYmKZMmeLS4gAAAAAgu2U6JCUnJ0uSwsPDtWvXLgUEBGRZUQAAAADgLk6vbhcdHZ0VdQAAAABAjuB0SJowYUKG28eMGXPPxQAAAACAuzkdkpYtW+bw/Pbt24qOjlaePHlUpkwZQhIAAACA+5rTIWnv3r2p2mJjY9W7d2916dLFJUUBAAAAgLs4/TlJaSlYsKDGjx+v0aNHu2I4AAAAAHAbl4QkSbp27ZquXbvmquEAAAAAwC2cvtxuxowZDs+NMTp79qwWLlyoNm3auKwwAAAAAHAHp0PSu+++6/Dcw8NDgYGB6tWrl0aNGuWywgAAAADAHficJAAAAACwcNk9SQAAAACQG2R6Jqlv376Z6vfhhx/eczEAAAAA4G6ZDkmRkZEKDQ1VrVq1ZIzJypoAAAAAwG0yHZKef/55LVq0SNHR0erTp4/++c9/qkiRIllZGwAAAABku0zfkzR79mydPXtWI0aM0FdffaWQkBA9+eSTWrNmDTNLAAAAAHINpxZu8Pb21lNPPaW1a9fqwIEDqlKligYOHKiwsDDFxcVlVY0AAAAAkG3ueXU7Dw8P2Ww2GWOUlJTkypoAAAAAwG2cCkkJCQlatGiRWrZsqfLly+vnn3/WrFmzdOrUKfn6+mZVjQAAAACQbTK9cMPAgQP12WefKSQkRH379tWiRYsUEBCQlbUBAAAAQLbLdEiaN2+eSpUqpdKlS2vjxo3auHFjmv2+/PJLlxUHAAAAANkt0yGpZ8+estlsWVkLAAAAALidUx8mCwAAAAC53T2vbgcAAAAAuREhCQAAAAAsCEkAAAAAYEFIAgAAAACLTIWk2rVr68qVK5KkCRMm6MaNG1laFAAAAAC4S6ZC0sGDBxUfHy9JGj9+vOLi4rK0KAAAAABwl0wtAV6zZk316dNHjzzyiIwxeuedd+Tr65tm3zFjxri0QAAAAADITpkKSZGRkRo7dqy+/vpr2Ww2rVq1SnnypN7VZrMRkgAAAADc1zIVkipUqKDPPvtMkuTh4aGoqCgVK1YsSwsDAAAAAHfIVEiySk5Ozoo6AAAAACBHcDokSdLx48c1bdo0HTx4UJJUuXJlvfjiiypTpoxLiwMAAACA7Ob05yStWbNGlStX1s6dO1W9enVVr15dO3bsUJUqVbR27dqsqBEAAAAAso3TM0mvvPKKhg4dqjfffDNV+8iRI9WyZUuXFQcAAAAA2c3pmaSDBw+qX79+qdr79u2rAwcOuKQoAAAAAHAXp0NSYGCg9u3bl6p93759rHgHAAAA4L7n9OV2/fv314ABA/Trr7/q4YcfliRt2bJFb731loYNG+byAgEAAAAgOzkdkkaPHi0/Pz9NmTJFo0aNkiQFBwdr3LhxGjx4sMsLBAAAAIDs5HRIstlsGjp0qIYOHarr169Lkvz8/FxeGAAAAAC4wz19TlIKwhEAAACA3MbphRsAAAAAIDcjJAEAAACABSEJAAAAACycCkm3b99W8+bNdfTo0ayqBwAAAADcyqmQlDdvXv30009ZVQsAAAAAuJ3Tl9v985//1AcffJAVtQAAAACA2zm9BPidO3f04Ycf6vvvv1edOnVUoEABh+1Tp051WXEAAAAAkN2cDkm//PKLateuLUk6cuSIwzabzeaaqgAAAADATZwOSevXr8+KOgAAAAAgR7jnJcCPHTumNWvW6ObNm5IkY4zLigIAAAAAd3E6JF26dEnNmzdX+fLl1bZtW509e1aS1K9fPw0fPtzlBQIAAABAdnI6JA0dOlR58+bVqVOnlD9/fnt7t27dtHr1apcWBwAAAADZzel7kr777jutWbNGDzzwgEN7uXLldPLkSZcVBgAAAADu4PRMUnx8vMMMUorLly/L29vbJUUBAAAAgLs4HZIaNWqkjz76yP7cZrMpOTlZkydPVtOmTV1aHAAAAABkN6cvt5s8ebKaN2+u3bt3KzExUSNGjND+/ft1+fJlbdmyJStqBAAAAIBs4/RMUtWqVXXkyBE98sgj6tSpk+Lj4/X4449r7969KlOmTFbUCAAAAADZxumZJEny9/fXq6++6upaAAAAAMDt7ikkXblyRR988IEOHjwoSapcubL69OmjIkWKuLQ4AAAAAMhuTl9ut2nTJoWFhWnGjBm6cuWKrly5ohkzZig8PFybNm3KihoBAAAAINs4PZMUERGhbt26ae7cufL09JQkJSUlaeDAgYqIiNDPP//s8iIBAAAAILs4PZN07NgxDR8+3B6QJMnT01PDhg3TsWPHXFocAAAAAGQ3p0NS7dq17fciWR08eFA1atRwSVEAAAAA4C6Zutzup59+sv978ODBevHFF3Xs2DE99NBDkqTt27dr9uzZevPNN7OmSgAAAADIJpkKSTVr1pTNZpMxxt42YsSIVP2efvppdevWzXXVAQAAAEA2y1RIio6Ozuo6AAAAACBHyFRICg0Nzeo6AAAAACBHuKcPkz1z5ow2b96sCxcuKDk52WHb4MGDXVIYAAAAALiD0yEpMjJS//rXv+Tl5aWiRYvKZrPZt9lsNkISAAAAgPua0yFp9OjRGjNmjEaNGiUPD6dXEAcAAACAHM3plHPjxg11796dgAQAAAAgV3I66fTr109LlizJiloAAAAAwO2cvtxu0qRJat++vVavXq1q1aopb968DtunTp3qsuIAAAAAILvdU0has2aNKlSoIEmpFm4AAAAAgPuZ0yFpypQp+vDDD9W7d+8sKAcAAAAA3Mvpe5K8vb3VsGHDrKjlrt58803ZbDYNGTLELccHAAAAkPs5HZJefPFFzZw5MytqydCuXbv03nvvqXr16tl+bAAAAAB/H05fbrdz506tW7dOX3/9tapUqZJq4YYvv/zSZcWliIuLU48ePfT+++/rtddec/n4AAAAAJDC6ZBUqFAhPf7441lRS7oiIiLUrl07tWjR4q4hKSEhQQkJCfbnsbGxWV0eAAAAgFzE6ZA0f/78rKgjXZ999pl++OEH7dq1K1P9J02apPHjx2dxVQAAAAByK6fvScpOp0+f1osvvqhPPvlE+fLly9Q+o0aN0rVr1+yP06dPZ3GVAAAAAHITp2eSwsPDM/w8pF9//fUvFWS1Z88eXbhwQbVr17a3JSUladOmTZo1a5YSEhLk6enpsI+3t7e8vb1dVgMAAACAvxenQ9Kfl9++ffu29u7dq9WrV+vll192VV2SpObNm+vnn392aOvTp48qVqyokSNHpgpIAAAAAPBXOR2SXnzxxTTbZ8+erd27d//lgqz8/PxUtWpVh7YCBQqoaNGiqdoBAAAAwBVcdk9SmzZttHTpUlcNBwAAAABu4fRMUnq++OILFSlSxFXDpWvDhg1ZfgwAAAAAf19Oh6RatWo5LNxgjNG5c+d08eJFzZkzx6XFAQAAAEB2czokde7c2eG5h4eHAgMD1aRJE1WsWNFVdQEAAACAWzgdksaOHZsVdQAAAABAjpCjP0wWAAAAALJbpmeSPDw8MvwQWUmy2Wy6c+fOXy4KAAAAANwl0yFp2bJl6W7btm2bZsyYoeTkZJcUBQAAAADukumQ1KlTp1Rthw8f1iuvvKKvvvpKPXr00IQJE1xaHAAAAABkt3u6J+nMmTPq37+/qlWrpjt37mjfvn1asGCBQkNDXV0fAAAAAGQrp0LStWvXNHLkSJUtW1b79+9XVFSUvvrqK1WtWjWr6gMAAACAbJXpy+0mT56st956S0FBQVq0aFGal98BAAAAwP0u0yHplVdekY+Pj8qWLasFCxZowYIFafb78ssvXVYcAAAAAGS3TIeknj173nUJcAAAAAC432U6JEVGRmZhGQAAAACQM9zT6nYAAAAAkFsRkgAAAADAgpAEAAAAABaEJAAAAACwICQBAAAAgAUhCQAAAAAsCEkAAAAAYEFIAgAAAAALQhIAAAAAWBCSAAAAAMCCkAQAAAAAFoQkAAAAALAgJAEAAACABSEJAAAAACwISQAAAABgQUgCAAAAAAtCEgAAAABYEJIAAAAAwIKQBAAAAAAWhCQAAAAAsCAkAQAAAIAFIQkAAAAALAhJAAAAAGBBSAIAAAAAC0ISAAAAAFgQkgAAAADAgpAEAAAAABaEJAAAAACwICQBAAAAgAUhCQAAAAAsCEkAAAAAYEFIAgAAAAALQhIAAAAAWBCSAAAAAMCCkAQAAAAAFoQkAAAAALAgJAEAAACABSEJAAAAACwISQAAAABgQUgCAAAAAAtCEgAAAABYEJIAAAAAwIKQBAAAAAAWhCQAAAAAsCAkAQAAAIAFIQkAAAAALAhJAAAAAGBBSAIAAAAAC0ISAAAAAFgQkgAAAADAgpAEAAAAABaEJAAAAACwICQBAAAAgAUhCQAAAAAsCEkAAAAAYEFIAgAAAAALQhIAAAAAWBCSAAAAAMCCkAQAAAAAFoQkAAAAALAgJAEAAACABSEJAAAAACwISQAAAABgQUgCAAAAAIscHZImTZqkunXrys/PT8WKFVPnzp11+PBhd5cFAAAAIBfL0SFp48aNioiI0Pbt27V27Vrdvn1bjz32mOLj491dGgAAAIBcKo+7C8jI6tWrHZ5HRkaqWLFi2rNnjx599FE3VQUAAAAgN8vRIenPrl27JkkqUqRIun0SEhKUkJBgfx4bG5vldQEAAADIPXL05XZWycnJGjJkiBo2bKiqVaum22/SpEny9/e3P0JCQrKxSgAAAAD3u/smJEVEROiXX37RZ599lmG/UaNG6dq1a/bH6dOns6lCAAAAALnBfXG53aBBg/T1119r06ZNeuCBBzLs6+3tLW9v72yqDAAAAEBuk6NDkjFGL7zwgpYtW6YNGzYoPDzc3SUBAAAAyOVydEiKiIjQp59+qhUrVsjPz0/nzp2TJPn7+8vHx8fN1QEAAADIjXL0PUlz587VtWvX1KRJE5UoUcL+WLx4sbtLAwAAAJBL5eiZJGOMu0sAAAAA8DeTo2eSAAAAACC7EZIAAAAAwIKQBAAAAAAWhCQAAAAAsCAkAQAAAIAFIQkAAAAALAhJAAAAAGBBSAIAAAAAC0ISAAAAAFgQkgAAAADAgpAEAAAAABaEJAAAAACwICQBAAAAgAUhCQAAAAAsCEkAAAAAYEFIAgAAAAALQhIAAAAAWBCSAAAAAMCCkAQAAAAAFoQkAAAAALAgJAEAAACABSEJAAAAACwISQAAAABgQUgCAAAAAAtCEgAAAABYEJIAAAAAwIKQBAAAAAAWhCQAAAAAsCAkAQAAAIAFIQkAAAAALAhJAAAAAGBBSAIAAAAAC0ISAAAAAFgQkgAAAADAgpAEAAAAABaEJAAAAACwyOPuAv5uTp06pZiYmCwbPyAgQKVKlcqy8QEAAIDcjpCUjU6dOqUKFSvp1s0bWXaMfD75dfjQQYISAAAAcI8ISdkoJiZGt27eUNH2w5W3aIjLx7996bQufT1FMTExhCQAAADgHhGS3CBv0RB5B5V1dxkAAAAA0sDCDQAAAABgQUgCAAAAAAtCEgAAAABYEJIAAAAAwIKQBAAAAAAWhCQAAAAAsCAkAQAAAIAFIQkAAAAALAhJAAAAAGBBSAIAAAAAC0ISAAAAAFgQkgAAAADAgpAEAAAAABaEJAAAAACwICQBAAAAgAUhCQAAAAAsCEkAAAAAYEFIAgAAAAALQhIAAAAAWBCSAAAAAMCCkAQAAAAAFoQkAAAAALAgJAEAAACABSEJAAAAACwISQAAAABgQUgCAAAAAAtCEgAAAABYEJIAAAAAwIKQBAAAAAAWhCQAAAAAsCAkAQAAAIAFIQkAAAAALAhJAAAAAGBBSAIAAAAAC0ISAAAAAFgQkgAAAADAgpAEAAAAABb3RUiaPXu2wsLClC9fPtWvX187d+50d0kAAAAAcqkcH5IWL16sYcOGaezYsfrhhx9Uo0YNtWrVShcuXHB3aQAAAAByoRwfkqZOnar+/furT58+qly5subNm6f8+fPrww8/dHdpAAAAAHKhPO4uICOJiYnas2ePRo0aZW/z8PBQixYttG3btjT3SUhIUEJCgv35tWvXJEmxsbFZW2wmxMXFSZISzh1TcuItl49/+/JvkqQ9e/bYj+VKHh4eSk5Odvm42TE+tbtn/Pt17Kwen9rdMz61u2d8anfP+NTunvHv19oPHz4sKev/Ro2Li3P73+QpxzfGZNgvR4ekmJgYJSUlqXjx4g7txYsX16FDh9LcZ9KkSRo/fnyq9pCQkCyp8V5cWTMrS8cfMGBAlo4PAACA3Cer/0Zt3Lhxlo7vjOvXr8vf3z/d7Tk6JN2LUaNGadiwYfbnycnJunz5sooWLSqbzXZPY8bGxiokJESnT59WwYIFXVUqkCHOO7gD5x3chXMP7sB59/djjNH169cVHBycYb8cHZICAgLk6emp8+fPO7SfP39eQUFBae7j7e0tb29vh7ZChQq5pJ6CBQvyDYRsx3kHd+C8g7tw7sEdOO/+XjKaQUqRoxdu8PLyUp06dRQVFWVvS05OVlRUlBo0aODGygAAAADkVjl6JkmShg0bpl69eunBBx9UvXr1NG3aNMXHx6tPnz7uLg0AAABALpTjQ1K3bt108eJFjRkzRufOnVPNmjW1evXqVIs5ZCVvb2+NHTs21WV8QFbivIM7cN7BXTj34A6cd0iPzdxt/TsAAAAA+BvJ0fckAQAAAEB2IyQBAAAAgAUhCQAAAAAsCEkAAAAAYEFIuovZs2crLCxM+fLlU/369bVz5053l4RcZNy4cbLZbA6PihUr2rffunVLERERKlq0qHx9ffXEE0+k+nBlIDM2bdqkDh06KDg4WDabTcuXL3fYbozRmDFjVKJECfn4+KhFixY6evSoQ5/Lly+rR48eKliwoAoVKqR+/fopLi4uG18F7jd3O+969+6d6mdg69atHfpw3sFZkyZNUt26deXn56dixYqpc+fOOnz4sEOfzPx+PXXqlNq1a6f8+fOrWLFievnll3Xnzp3sfClwI0JSBhYvXqxhw4Zp7Nix+uGHH1SjRg21atVKFy5ccHdpyEWqVKmis2fP2h+bN2+2bxs6dKi++uorLVmyRBs3btSZM2f0+OOPu7Fa3K/i4+NVo0YNzZ49O83tkydP1owZMzRv3jzt2LFDBQoUUKtWrXTr1i17nx49emj//v1au3atvv76a23atEkDBgzIrpeA+9DdzjtJat26tcPPwEWLFjls57yDszZu3KiIiAht375da9eu1e3bt/XYY48pPj7e3uduv1+TkpLUrl07JSYmauvWrVqwYIEiIyM1ZswYd7wkuINBuurVq2ciIiLsz5OSkkxwcLCZNGmSG6tCbjJ27FhTo0aNNLddvXrV5M2b1yxZssTedvDgQSPJbNu2LZsqRG4kySxbtsz+PDk52QQFBZm3337b3nb16lXj7e1tFi1aZIwx5sCBA0aS2bVrl73PqlWrjM1mM7///nu21Y7715/PO2OM6dWrl+nUqVO6+3DewRUuXLhgJJmNGzcaYzL3+/Xbb781Hh4e5ty5c/Y+c+fONQULFjQJCQnZ+wLgFswkpSMxMVF79uxRixYt7G0eHh5q0aKFtm3b5sbKkNscPXpUwcHBKl26tHr06KFTp05Jkvbs2aPbt287nIMVK1ZUqVKlOAfhUtHR0Tp37pzDuebv76/69evbz7Vt27apUKFCevDBB+19WrRoIQ8PD+3YsSPba0busWHDBhUrVkwVKlTQ888/r0uXLtm3cd7BFa5duyZJKlKkiKTM/X7dtm2bqlWrpuLFi9v7tGrVSrGxsdq/f382Vg93ISSlIyYmRklJSQ7fHJJUvHhxnTt3zk1VIbepX7++IiMjtXr1as2dO1fR0dFq1KiRrl+/rnPnzsnLy0uFChVy2IdzEK6Wcj5l9PPu3LlzKlasmMP2PHnyqEiRIpyPuGetW7fWRx99pKioKL311lvauHGj2rRpo6SkJEmcd/jrkpOTNWTIEDVs2FBVq1aVpEz9fj137lyaPxNTtiH3y+PuAoC/szZt2tj/Xb16ddWvX1+hoaH6/PPP5ePj48bKACDrde/e3f7vatWqqXr16ipTpow2bNig5s2bu7Ey5BYRERH65ZdfHO73BTKDmaR0BAQEyNPTM9VKJ+fPn1dQUJCbqkJuV6hQIZUvX17Hjh1TUFCQEhMTdfXqVYc+nINwtZTzKaOfd0FBQakWrblz544uX77M+QiXKV26tAICAnTs2DFJnHf4awYNGqSvv/5a69ev1wMPPGBvz8zv16CgoDR/JqZsQ+5HSEqHl5eX6tSpo6ioKHtbcnKyoqKi1KBBAzdWhtwsLi5Ox48fV4kSJVSnTh3lzZvX4Rw8fPiwTp06xTkIlwoPD1dQUJDDuRYbG6sdO3bYz7UGDRro6tWr2rNnj73PunXrlJycrPr162d7zcidfvvtN126dEklSpSQxHmHe2OM0aBBg7Rs2TKtW7dO4eHhDtsz8/u1QYMG+vnnnx1C+tq1a1WwYEFVrlw5e14I3MvdK0fkZJ999pnx9vY2kZGR5sCBA2bAgAGmUKFCDiudAH/F8OHDzYYNG0x0dLTZsmWLadGihQkICDAXLlwwxhjz3HPPmVKlSpl169aZ3bt3mwYNGpgGDRq4uWrcj65fv2727t1r9u7daySZqVOnmr1795qTJ08aY4x58803TaFChcyKFSvMTz/9ZDp16mTCw8PNzZs37WO0bt3a1KpVy+zYscNs3rzZlCtXzjz11FPuekm4D2R03l2/ft289NJLZtu2bSY6Otp8//33pnbt2qZcuXLm1q1b9jE47+Cs559/3vj7+5sNGzaYs2fP2h83btyw97nb79c7d+6YqlWrmscee8zs27fPrF692gQGBppRo0a54yXBDQhJdzFz5kxTqlQp4+XlZerVq2e2b9/u7pKQi3Tr1s2UKFHCeHl5mZIlS5pu3bqZY8eO2bffvHnTDBw40BQuXNjkz5/fdOnSxZw9e9aNFeN+tX79eiMp1aNXr17GmD+WAR89erQpXry48fb2Ns2bNzeHDx92GOPSpUvmqaeeMr6+vqZgwYKmT58+5vr16254NbhfZHTe3bhxwzz22GMmMDDQ5M2b14SGhpr+/fun+o9Izjs4K61zTpKZP3++vU9mfr+eOHHCtGnTxvj4+JiAgAAzfPhwc/v27Wx+NXAXmzHGZPfsFQAAAADkVNyTBAAAAAAWhCQAAAAAsCAkAQAAAIAFIQkAAAAALAhJAAAAAGBBSAIAAAAAC0ISAAAAAFgQkgAAAADAgpAEAMjxxo0bp5o1a9qf9+7dW507d3ZbPQCA3I2QBAC4J6dPn1bfvn0VHBwsLy8vhYaG6sUXX9SlS5ey/NjTp09XZGSk/XmTJk00ZMiQvzzujRs3NGrUKJUpU0b58uVTYGCgGjdurBUrVvzlsQEA94887i4AAHD/+fXXX9WgQQOVL19eixYtUnh4uPbv36+XX35Zq1at0vbt21WkSJEsO76/v3+WjPvcc89px44dmjlzpipXrqxLly5p69atWRr8EhMT5eXllWXjAwCcx0wSAMBpERER8vLy0nfffafGjRurVKlSatOmjb7//nv9/vvvevXVV+19bTabli9f7rB/oUKFHGaCRo4cqfLlyyt//vwqXbq0Ro8erdu3b6d7fOvldr1799bGjRs1ffp02Ww22Ww2RUdHq2zZsnrnnXcc9tu3b59sNpuOHTuW5rgrV67Uv//9b7Vt21ZhYWGqU6eOXnjhBfXt29feJyEhQSNHjlRISIi8vb1VtmxZffDBB/btGzduVL169eTt7a0SJUrolVde0Z07d+zbmzRpokGDBmnIkCEKCAhQq1atJEm//PKL2rRpI19fXxUvXlzPPPOMYmJi0n0PAABZh5AEAHDK5cuXtWbNGg0cOFA+Pj4O24KCgtSjRw8tXrxYxphMj+nn56fIyEgdOHBA06dP1/vvv6933303U/tOnz5dDRo0UP/+/XX27FmdPXtWpUqVUt++fTV//nyHvvPnz9ejjz6qsmXLpjlWUFCQvv32W12/fj3d4/Xs2VOLFi3SjBkzdPDgQb333nvy9fWVJP3+++9q27at6tatqx9//FFz587VBx98oNdee81hjAULFsjLy0tbtmzRvHnzdPXqVTVr1ky1atXS7t27tXr1ap0/f15PPvlkpt4DAIBrcbkdAMApR48elTFGlSpVSnN7pUqVdOXKFV28eFHFihXL1Jj/+c9/7P8OCwvTSy+9pM8++0wjRoy4677+/v7y8vJS/vz5FRQUZG/v3bu3xowZo507d6pevXq6ffu2Pv3001SzS1b//e9/1aNHDxUtWlQ1atTQI488oq5du6phw4aSpCNHjujzzz/X2rVr1aJFC0lS6dKl7fvPmTNHISEhmjVrlmw2mypWrKgzZ85o5MiRGjNmjDw8/vi/yXLlymny5Mn2/V577TXVqlVLb7zxhr3tww8/VEhIiI4cOaLy5cvf9X0AALgOM0kAgHtyt5kiZ+6zWbx4sRo2bKigoCD5+vrqP//5j06dOvWX6gsODla7du304YcfSpK++uorJSQk6B//+Ee6+zz66KP69ddfFRUVpa5du2r//v1q1KiRJk6cKOmPy/U8PT3VuHHjNPc/ePCgGjRoIJvNZm9r2LCh4uLi9Ntvv9nb6tSp47Dfjz/+qPXr18vX19f+qFixoiTp+PHj9/YGAADuGSEJAOCUsmXLymaz6eDBg2luP3jwoAIDA1WoUCFJf9yT9OdAZb3faNu2berRo4fatm2rr7/+Wnv37tWrr76qxMTEv1zrs88+q88++0w3b97U/Pnz1a1bN+XPnz/DffLmzatGjRpp5MiR+u677zRhwgRNnDhRiYmJqS4vvFcFChRweB4XF6cOHTpo3759Do+jR4/q0UcfdckxAQCZx+V2AACnFC1aVC1bttScOXM0dOhQh+Bw7tw5ffLJJ4qIiLC3BQYG6uzZs/bnR48e1Y0bN+zPt27dqtDQUIfFHk6ePOlUTV5eXkpKSkrV3rZtWxUoUEBz587V6tWrtWnTJqfGlaTKlSvrzp07unXrlqpVq6bk5GRt3LjRfrmdVaVKlbR06VIZY+yzSVu2bJGfn58eeOCBdI9Ru3ZtLV26VGFhYcqTh1/NAOBuzCQBAJw2a9YsJSQkqFWrVtq0aZNOnz6t1atXq2XLlipfvrzGjBlj79usWTPNmjVLe/fu1e7du/Xcc88pb9689u3lypXTqVOn9Nlnn+n48eOaMWOGli1b5lQ9YWFh2rFjh06cOKGYmBglJydLkjw9PdW7d2+NGjVK5cqVU4MGDTIcp0mTJnrvvfe0Z88enThxQt9++63+/e9/q2nTpipYsKDCwsLUq1cv9e3bV8uXL1d0dLQ2bNigzz//XJI0cOBAnT59Wi+88IIOHTqkFStWaOzYsRo2bJj9fqS0RERE6PLly3rqqae0a9cuHT9+XGvWrFGfPn3SDH8AgKxFSAIAOK1cuXLatWuXSpcurSeffFKhoaFq06aNypcvry1btthXe5OkKVOmKCQkRI0aNdLTTz+tl156yeGSt44dO2ro0KEaNGiQatasqa1bt2r06NFO1fPSSy/J09NTlStXVmBgoMP9TP369VNiYqL69Olz13FatWqlBQsW6LHHHlOlSpX0wgsvqFWrVvYQJElz585V165dNXDgQFWsWFH9+/dXfHy8JKlkyZL69ttvtXPnTtWoUUPPPfec+vXr57AwRVqCg4O1ZcsWJSUl6bHHHlO1atU0ZMgQFSpUKMNwBQDIGjbjzBqtAACkY+zYsZo6darWrl2rhx56yN3l2P3vf/9T8+bNdfr0aRUvXtzd5QAA7gOEJACAy8yfP1/Xrl3T4MGD3T4DkpCQoIsXL6pXr14KCgrSJ5984tZ6AAD3D0ISACBXioyMVL9+/VSzZk2tXLlSJUuWdHdJAID7BCEJAAAAACy4GxQAAAAALAhJAAAAAGBBSAIAAAAAC0ISAAAAAFgQkgAAAADAgpAEAAAAABaEJAAAAACwICQBAAAAgMX/A3Cvl1dAYD1cAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 1000x600 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#visualization showing distribution of quality scores in SARS2 dataFrame\n",
    "plt.figure(figsize=(10, 6))\n",
    "plt.hist(SARS2_df['QUAL'], bins=30, edgecolor='black')\n",
    "plt.title('Distribution of Quality Scores (SARS2)')\n",
    "plt.xlabel('Quality Score')\n",
    "plt.ylabel('Number of Mutations')\n",
    "plt.savefig('SARS2_quality_score_distribution.png')\n",
    "plt.show()\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3",
   "name": "python3"
  },
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
